// tslint-sonarts/lib/rules/cognitiveComplexityRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
const cognitiveComplexity_1 = require("../utils/cognitiveComplexity");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    get threshold() {
        if (this.ruleArguments[0] !== undefined) {
            return this.ruleArguments[0];
        }
        return Rule.DEFAULT_THRESHOLD;
    }
    apply(sourceFile) {
        const visitor = new sonarAnalysis_1.SonarRuleVisitor(this.getOptions().ruleName);
        const functionCollector = new cognitiveComplexity_1.FunctionCollector();
        functionCollector.visit(sourceFile);
        functionCollector.functionComplexities.forEach(functionComplexity => {
            if (functionComplexity.complexity > this.threshold) {
                const issue = visitor.addIssue(navigation_1.functionLikeMainToken(functionComplexity.functionNode), getMessage(functionComplexity.functionNode, functionComplexity.complexity, this.threshold));
                issue.setCost(functionComplexity.complexity - this.threshold);
                functionComplexity.nodes.forEach(complexityNode => issue.addSecondaryLocation(sonarAnalysis_1.getIssueLocationAtNode(complexityNode.node, secondaryMessage(complexityNode.complexity))));
            }
        });
        return visitor.getIssues();
        function secondaryMessage(complexity) {
            if (complexity > 1) {
                return `+${complexity} (incl. ${complexity - 1} for nesting)`;
            }
            else {
                return "+" + complexity;
            }
        }
    }
}
Rule.metadata = {
    ruleName: "cognitive-complexity",
    description: "Cognitive Complexity of functions should not be too high",
    rationale: tslint.Utils.dedent `
      Cognitive Complexity is a measure of how hard the control flow of a function is to understand.
      Functions with high Cognitive Complexity will be difficult to maintain.`,
    options: { type: "number" },
    optionsDescription: `The maximum authorized complexity. Default is ${Rule.DEFAULT_THRESHOLD}.`,
    optionExamples: [[true, 10]],
    rspecKey: "RSPEC-3776",
    type: "maintainability",
    typescriptOnly: false,
};
Rule.DEFAULT_THRESHOLD = 15;
exports.Rule = Rule;
function getMessage(node, complexity, threshold) {
    const functionName = getFunctionName(node);
    return `Refactor this ${functionName} to reduce its Cognitive Complexity from ${complexity} to the ${threshold} allowed.`;
}
function getFunctionName(node) {
    switch (node.kind) {
        case ts.SyntaxKind.MethodDeclaration:
            return "method";
        case ts.SyntaxKind.Constructor:
            return "constructor";
        case ts.SyntaxKind.GetAccessor:
        case ts.SyntaxKind.SetAccessor:
            return "accessor";
        default:
            return "function";
    }
}
