// tslint-sonarts/lib/rules/consecutiveOverloadsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    applyWithProgram(sourceFile, program) {
        return new Visitor(this.getOptions().ruleName, program).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "consecutive-overloads",
    description: "Method overloads should be grouped together",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4136",
    type: "maintainability",
    typescriptOnly: true,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    visitTypeLiteral(node) {
        this.checkOverloads(node.members);
        super.visitTypeLiteral(node);
    }
    visitInterfaceDeclaration(node) {
        this.checkOverloads(node.members);
        super.visitInterfaceDeclaration(node);
    }
    visitClassDeclaration(node) {
        this.checkOverloads(node.members);
        super.visitClassDeclaration(node);
    }
    visitModuleDeclaration(node) {
        const moduleBody = node.body;
        if (nodes_1.is(moduleBody, ts.SyntaxKind.ModuleBlock)) {
            const functionLike = moduleBody.statements.filter(stmt => nodes_1.isSignatureDeclaration(stmt));
            this.checkOverloads(functionLike);
        }
        super.visitModuleDeclaration(node);
    }
    visitSourceFile(node) {
        this.checkOverloads(node.statements.filter(stmt => nodes_1.isSignatureDeclaration(stmt)));
        super.visitSourceFile(node);
    }
    checkOverloads(nodes) {
        const misplacedOverloads = getMisplacedOverloads(nodes);
        for (const [first, ...rest] of misplacedOverloads.values()) {
            if (rest.length > 0) {
                const issue = this.addIssue(first, `All '${printOverload(first)}' signatures should be adjacent`);
                rest.forEach(n => issue.addSecondaryLocation(n, "Non-adjacent overload"));
            }
        }
    }
}
function getMisplacedOverloads(overloads) {
    const result = new Map();
    let lastKey;
    for (const node of overloads) {
        if (node.kind === ts.SyntaxKind.SemicolonClassElement || isAccessor(node)) {
            continue;
        }
        const key = getOverloadKey(node);
        if (key !== undefined) {
            const overloads = result.get(key);
            if (overloads && lastKey !== key) {
                overloads.push(node);
            }
            if (overloads === undefined) {
                result.set(key, [node]);
            }
            lastKey = key;
        }
        else {
            lastKey = undefined;
        }
    }
    return result;
}
function isAccessor(member) {
    return member.kind === ts.SyntaxKind.GetAccessor || member.kind === ts.SyntaxKind.SetAccessor;
}
function printOverload(node) {
    const info = getOverloadInfo(node);
    return typeof info === "string" ? info : info === undefined ? "<unknown>" : info.name;
}
function getOverloadKey(node) {
    const info = getOverloadInfo(node);
    if (info === undefined) {
        return undefined;
    }
    const [computed, name] = typeof info === "string" ? [false, info] : [info.computed, info.name];
    const isStatic = node.modifiers && node.modifiers.map(m => m.kind).indexOf(ts.SyntaxKind.StaticKeyword) !== -1;
    return (computed ? "0" : "1") + (isStatic ? "0" : "1") + name;
}
function getOverloadInfo(node) {
    switch (node.kind) {
        case ts.SyntaxKind.ConstructSignature:
        case ts.SyntaxKind.Constructor:
            return "constructor";
        case ts.SyntaxKind.CallSignature:
            return "()";
        default: {
            if (!nodes_1.isSignatureDeclaration(node) || node.name === undefined) {
                return undefined;
            }
            const { name } = node;
            switch (name.kind) {
                case ts.SyntaxKind.Identifier:
                    return name.text;
                case ts.SyntaxKind.ComputedPropertyName:
                    const { expression } = name;
                    return nodes_1.isLiteralExpression(expression) ? expression.text : { name: expression.getText(), computed: true };
                default:
                    return nodes_1.isLiteralExpression(name) ? name.text : undefined;
            }
        }
    }
}
