// tslint-sonarts/lib/rules/maxUnionSizeRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.AbstractRule {
    get max() {
        return this.ruleArguments[0] || Rule.DEFAULT_MAX;
    }
    static message(max) {
        return `Refactor this union type to have less than ${max} elements.`;
    }
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName, this.max).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "max-union-size",
    description: "Union types should not have too many elements",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "Maximum elements authorized in a union type definition.",
    options: { type: "number" },
    rspecKey: "RSPEC-4622",
    type: "maintainability",
    typescriptOnly: true,
    optionExamples: [[true, 5]],
};
Rule.DEFAULT_MAX = 3;
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor(ruleName, max) {
        super(ruleName);
        this.max = max;
    }
    visitUnionTypeNode(node) {
        if (node.types.length > this.max && !this.isFromTypeStatement(node)) {
            this.addIssue(node, Rule.message(this.max));
        }
        super.visitUnionTypeNode(node);
    }
    isFromTypeStatement(node) {
        return nodes_1.is(node.parent, ts.SyntaxKind.TypeAliasDeclaration);
    }
}
