// tslint-sonarts/lib/rules/mccabeComplexityRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const navigation_1 = require("../utils/navigation");
const cyclomaticComplexity_1 = require("../utils/cyclomaticComplexity");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    get threshold() {
        if (this.ruleArguments[0] !== undefined) {
            return this.ruleArguments[0];
        }
        return Rule.DEFAULT_THRESHOLD;
    }
    static message(complexity, threshold) {
        return `The Cyclomatic Complexity of this function is ${complexity} which is greater than ${threshold} authorized.`;
    }
    apply(sourceFile) {
        return new Visitor(this.getOptions(), this.threshold).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "mccabe-complexity",
    description: "Functions should not be too complex",
    rspecKey: "RSPEC-1541",
    type: "maintainability",
    typescriptOnly: false,
    optionsDescription: `The maximum authorized complexity can be provided. Default is ${Rule.DEFAULT_THRESHOLD}.`,
    options: { type: "number" },
    optionExamples: [true, [true, 15]],
};
Rule.DEFAULT_THRESHOLD = 10;
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor(options, threshold) {
        super(options.ruleName);
        this.threshold = threshold;
    }
    visitFunctionLikeDeclaration(node) {
        const complexityNodes = cyclomaticComplexity_1.getFunctionComplexityNodes(node);
        const functionComplexity = complexityNodes.length;
        if (functionComplexity > this.threshold) {
            const issue = this.addIssue(navigation_1.functionLikeMainToken(node), Rule.message(functionComplexity, this.threshold));
            issue.setCost(functionComplexity - this.threshold);
            complexityNodes.forEach(node => issue.addSecondaryLocation(sonarAnalysis_1.getIssueLocationAtNode(node, "+1")));
        }
        super.visitFunctionLikeDeclaration(node);
    }
}
