// tslint-sonarts/lib/rules/noAccessorFieldMismatchRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const nav = require("../utils/navigation");
const tslint = require("tslint");
const ts = require("typescript");
const builder_1 = require("../symbols/builder");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.TypedRule {
    applyWithProgram(sourceFile, program) {
        const symbols = builder_1.SymbolTableBuilder.build(sourceFile, program);
        return new Visitor(this.getOptions(), symbols).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-accessor-field-mismatch",
    description: "Getters and setters should access the expected fields",
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4275",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor(options, symbols) {
        super(options.ruleName);
        this.symbols = symbols;
    }
    visitMethodDeclaration(method) {
        this.visitAccessor(method, this.setterOrGetter(method));
        super.visitMethodDeclaration(method);
    }
    visitSetAccessor(accessor) {
        this.visitAccessor(accessor, { type: "setter", name: Visitor.getName(accessor) });
        super.visitSetAccessor(accessor);
    }
    visitGetAccessor(accessor) {
        this.visitAccessor(accessor, { type: "getter", name: Visitor.getName(accessor) });
        super.visitGetAccessor(accessor);
    }
    visitAccessor(accessor, setterOrGetter) {
        if (!setterOrGetter) {
            return;
        }
        const containingStructure = nav.firstAncestor(accessor, [
            ts.SyntaxKind.ClassDeclaration,
            ts.SyntaxKind.ClassExpression,
            ts.SyntaxKind.ObjectLiteralExpression,
        ]);
        let matchingFields;
        let accessorIsPublic;
        if (containingStructure.kind === ts.SyntaxKind.ObjectLiteralExpression) {
            matchingFields = Visitor.matchingFields(Array.from(containingStructure.properties), [], setterOrGetter.name);
            accessorIsPublic = true;
        }
        else {
            matchingFields = Visitor.matchingFields(Array.from(containingStructure.members), Visitor.fieldsDeclaredInConstructorParameters(containingStructure), setterOrGetter.name);
            accessorIsPublic = Visitor.isPublic(accessor);
        }
        if (accessorIsPublic &&
            accessor.body &&
            matchingFields.length > 0 &&
            Visitor.bodyIsSimple(accessor.body, setterOrGetter.type) &&
            !this.fieldIsUsed(accessor, matchingFields)) {
            this.addIssue(accessor.name, `Refactor this ${setterOrGetter.type} so that it actually refers to the property '${matchingFields[0]
                .name.getText()}'`).addSecondaryLocation(matchingFields[0], "Property which should be referred.");
        }
    }
    static isPublic(method) {
        const modifier = nav.accessModifier(method);
        return !!modifier && modifier.kind === ts.SyntaxKind.PublicKeyword;
    }
    static getName(accessor) {
        return accessor.name.getText().toLowerCase();
    }
    setterOrGetter(method) {
        const name = method.name.getText().toLowerCase();
        if (name.startsWith("set") || name.startsWith("Set")) {
            if (method.parameters.length === 1) {
                return { type: "setter", name: name.substring(3) };
            }
        }
        if (name.startsWith("get") || name.startsWith("Get")) {
            if (method.parameters.length === 0) {
                return { type: "getter", name: name.substring(3) };
            }
        }
    }
    static matchingFields(members, constructorDeclaredParameters, targetName) {
        return members
            .filter(element => element.kind === ts.SyntaxKind.PropertyDeclaration ||
            element.kind === ts.SyntaxKind.PropertyAssignment ||
            element.kind === ts.SyntaxKind.ShorthandPropertyAssignment ||
            element.kind === ts.SyntaxKind.JsxAttribute)
            .map(element => element)
            .concat(constructorDeclaredParameters)
            .filter(element => !!element.name && Visitor.fieldNameMatches(element.name.getText(), targetName));
    }
    static fieldNameMatches(fieldName, targetName) {
        const fieldNameLowerCase = fieldName.toLowerCase();
        const underscoredTargetName = "_" + targetName;
        return fieldNameLowerCase === targetName || fieldNameLowerCase === underscoredTargetName;
    }
    static fieldsDeclaredInConstructorParameters(containingClass) {
        const constr = nav.constructorOf(containingClass);
        if (constr) {
            return constr.parameters.filter(parameter => nav.accessModifier(parameter) || nav.isReadonly(parameter));
        }
        else {
            return [];
        }
    }
    static bodyIsSimple(body, type) {
        if (body.statements.length > 1) {
            return false;
        }
        if (body.statements.length == 0) {
            return true;
        }
        const statement = body.statements[0];
        if (type === "setter") {
            if (statement.kind === ts.SyntaxKind.ExpressionStatement) {
                return statement.expression.kind === ts.SyntaxKind.BinaryExpression;
            }
            return false;
        }
        else {
            if (statement.kind === ts.SyntaxKind.ReturnStatement) {
                const expression = statement.expression;
                if (expression) {
                    return !!expression.getText().startsWith("this.");
                }
            }
            return false;
        }
    }
    fieldIsUsed(method, fields) {
        const body = method.body;
        for (const field of fields) {
            const usage = this.symbols.getUsage(field.name);
            if (usage && body && !!this.symbols.allUsages(usage.symbol).find(usage => usage.isUsedInside(body))) {
                return true;
            }
        }
        return false;
    }
}
