// tslint-sonarts/lib/rules/noBigFunctionRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const navigation_1 = require("../utils/navigation");
class Rule extends tslint.Rules.AbstractRule {
    get max() {
        return this.ruleArguments[0] || Rule.DEFAULT_MAX;
    }
    static message(functionSize, max) {
        return `This function has ${functionSize} lines, which is greater than the ${max} lines authorized. Split it into smaller functions.`;
    }
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName, this.max).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-big-function",
    description: "Functions should not have too many lines of code",
    rationale: tslint.Utils.dedent ``,
    rspecKey: "RSPEC-138",
    type: "maintainability",
    typescriptOnly: false,
    options: { type: "number" },
    optionsDescription: `Maximum authorized lines of code in a function. Default is ${Rule.DEFAULT_MAX}.`,
    optionExamples: [[true, 100]],
};
Rule.DEFAULT_MAX = 200;
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor(ruleName, max) {
        super(ruleName);
        this.max = max;
    }
    visitFunctionLikeDeclaration(node) {
        if (node.body) {
            const startLine = navigation_1.startLineAndCharacter(node.body).line;
            const endLine = navigation_1.endLineAndCharacter(node.body).line;
            const functionSize = endLine - startLine;
            if (functionSize > this.max) {
                this.addIssue(navigation_1.functionLikeMainToken(node), Rule.message(functionSize, this.max));
            }
        }
        super.visitFunctionLikeDeclaration(node);
    }
}
