// tslint-sonarts/lib/rules/noDuplicateStringRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    get threshold() {
        return this.ruleArguments[0] || Rule.DEFAULT_THRESHOLD;
    }
    static message(times) {
        return `Define a constant instead of duplicating this literal ${times} times.`;
    }
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName, this.threshold).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-duplicate-string",
    description: "String literals should not be duplicated",
    rationale: tslint.Utils.dedent ``,
    rspecKey: "RSPEC-1192",
    type: "maintainability",
    typescriptOnly: false,
    optionsDescription: `Number of times a literal must be duplicated to trigger an issue. Default is ${Rule.DEFAULT_THRESHOLD}.`,
    options: { type: "number" },
    optionExamples: [true, [true, 5]],
};
Rule.DEFAULT_THRESHOLD = 3;
Rule.MIN_LENGTH = 10;
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor(ruleName, threshold) {
        super(ruleName);
        this.threshold = threshold;
        this.literals = new Map();
    }
    visitStringLiteral(node) {
        const stringContent = getStringContent(node);
        if (stringContent.length >= Rule.MIN_LENGTH &&
            !stringContent.match(Visitor.noSeparatorRegexp) &&
            !jsxAttribute(node)) {
            const sameStringLiterals = this.literals.get(stringContent) || [];
            sameStringLiterals.push(node);
            this.literals.set(stringContent, sameStringLiterals);
        }
        super.visitStringLiteral(node);
    }
    visitSourceFile(node) {
        super.visitSourceFile(node);
        this.literals.forEach(literals => {
            if (literals.length >= this.threshold) {
                const issue = this.addIssue(literals[0], Rule.message(literals.length)).setCost(literals.length - 1);
                literals.slice(1).forEach(literal => issue.addSecondaryLocation(literal, "Duplicate"));
            }
        });
    }
}
Visitor.noSeparatorRegexp = /^\w*$/;
function getStringContent(node) {
    return node.getText().substr(1, node.getText().length - 2);
}
function jsxAttribute(node) {
    const parent = node.parent;
    return parent && parent.kind === ts.SyntaxKind.JsxAttribute;
}
