// tslint-sonarts/lib/rules/noDuplicateInCompositeRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const areEquivalent_1 = require("../utils/areEquivalent");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-duplicate-in-composite",
    description: "Union and intersection types should not be defined with duplicated elements",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4621",
    type: "maintainability",
    typescriptOnly: true,
};
Rule.MESSAGE = "Remove this duplicated type or replace with another one.";
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitIntersectionTypeNode(node) {
        this.checkDuplication(node);
        super.visitIntersectionTypeNode(node);
    }
    visitUnionTypeNode(node) {
        this.checkDuplication(node);
        super.visitUnionTypeNode(node);
    }
    checkDuplication(node) {
        const alreadyChecked = {};
        node.types.forEach((currentType, currentIndex) => {
            if (alreadyChecked[currentIndex]) {
                return;
            }
            const duplicateTypes = [];
            node.types.forEach((type, index) => {
                if (areEquivalent_1.default(type, currentType)) {
                    alreadyChecked[index] = true;
                    duplicateTypes.push(type);
                }
            });
            if (duplicateTypes.length > 1) {
                const issue = this.addIssue(duplicateTypes[1], Rule.MESSAGE);
                issue.addSecondaryLocation(duplicateTypes[0], "Original");
                duplicateTypes.slice(2).forEach(type => issue.addSecondaryLocation(type, "Another duplicate"));
            }
        });
    }
}
