// tslint-sonarts/lib/rules/noDuplicatedBranchesRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const areEquivalent_1 = require("../utils/areEquivalent");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.AbstractRule {
    static formatMessage(type, line) {
        return `This ${type}'s code block is the same as the block for the ${type} on line ${line}.`;
    }
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-duplicated-branches",
    description: "Two branches in a conditional structure should not have exactly the same implementation",
    rationale: tslint.Utils.dedent `
      Having two cases in a switch statement or two branches in an if chain with the same implementation is at best
      duplicate code, and at worst a coding error. If the same logic is truly needed for both instances, then in an if
      chain they should be combined, or for a switch, one should fall through to the other.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-1871",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitIfStatement(node) {
        // don't visit `else if` statements
        if (!node.parent || node.parent.kind !== ts.SyntaxKind.IfStatement) {
            const branches = this.collectIfBranches(node);
            for (let i = 1; i < branches.length; i++) {
                if (this.hasRequiredSize(branches[i])) {
                    for (let j = 0; j < i; j++) {
                        if (this.compareIfBranches(branches[i], branches[j])) {
                            break;
                        }
                    }
                }
            }
        }
        super.visitIfStatement(node);
    }
    visitSwitchStatement(node) {
        const { clauses } = node.caseBlock;
        for (let i = 1; i < clauses.length; i++) {
            const firstClauseWithoutBreak = this.takeWithoutBreak(this.expandSingleBlockStatement(Array.from(clauses[i].statements)));
            if (this.hasRequiredSize(firstClauseWithoutBreak)) {
                for (let j = 0; j < i; j++) {
                    const secondClauseWithoutBreak = this.takeWithoutBreak(this.expandSingleBlockStatement(Array.from(clauses[j].statements)));
                    if (areEquivalent_1.default(firstClauseWithoutBreak, secondClauseWithoutBreak)) {
                        this.addIssue(clauses[i], Rule.formatMessage("case", this.getLine(clauses[j]))).addSecondaryLocation(sonarAnalysis_1.getIssueLocationAtNode(clauses[j]));
                        break;
                    }
                }
            }
        }
        super.visitSwitchStatement(node);
    }
    compareIfBranches(a, b) {
        const equivalent = areEquivalent_1.default(a, b);
        if (equivalent) {
            this.addIssue(a, Rule.formatMessage("branch", this.getLine(b))).addSecondaryLocation(sonarAnalysis_1.getIssueLocationAtNode(b));
        }
        return equivalent;
    }
    hasRequiredSize(node) {
        const nodes = Array.isArray(node) ? node : node.getChildren();
        const children = nodes.filter(child => child.kind !== ts.SyntaxKind.OpenBraceToken && child.kind !== ts.SyntaxKind.CloseBraceToken);
        return children.length > 0 && this.getLastLine(children[children.length - 1]) > this.getLine(children[0]);
    }
    collectIfBranches(node) {
        const branches = [node.thenStatement];
        let statement = node.elseStatement;
        while (statement) {
            if (nodes_1.isIfStatement(statement)) {
                branches.push(statement.thenStatement);
                statement = statement.elseStatement;
            }
            else {
                branches.push(statement);
                break;
            }
        }
        return branches;
    }
    getLine(node) {
        return node.getSourceFile().getLineAndCharacterOfPosition(node.getStart()).line + 1;
    }
    getLastLine(node) {
        return node.getSourceFile().getLineAndCharacterOfPosition(node.getEnd()).line + 1;
    }
    takeWithoutBreak(nodes) {
        return nodes.length > 0 && nodes[nodes.length - 1].kind === ts.SyntaxKind.BreakStatement
            ? nodes.slice(0, -1)
            : nodes;
    }
    expandSingleBlockStatement(nodes) {
        return nodes.length === 1 && nodes_1.isBlock(nodes[0]) ? Array.from(nodes[0].statements) : nodes;
    }
}
