// tslint-sonarts/lib/rules/noElementOverwriteRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const Lint = require("tslint");
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
const sonarUtils_1 = require("../runner/sonarUtils");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
const areEquivalent_1 = require("../utils/areEquivalent");
class Rule extends Lint.Rules.TypedRule {
    applyWithProgram(sourceFile, program) {
        return new Visitor(this.getOptions().ruleName, program).visit(sourceFile).getIssues();
    }
    static message(index, previousUsage) {
        const line = sonarUtils_1.nodeToSonarLine(previousUsage);
        return `Verify this is the index that was intended; "${index}" was already set on line ${line}.`;
    }
}
Rule.metadata = {
    description: "Collection elements should not be replaced unconditionally",
    options: null,
    optionsDescription: "",
    rationale: Lint.Utils.dedent `
      It is highly suspicious when a value is saved for a key or index and then unconditionally overwritten. Such overwrites are likely an error.`,
    rspecKey: "RSPEC-4143",
    ruleName: "no-element-overwrite",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    visitSourceFile(node) {
        this.checkStatements(node.statements);
        super.visitSourceFile(node);
    }
    visitBlock(node) {
        this.checkStatements(node.statements);
        super.visitBlock(node);
    }
    visitCaseClause(node) {
        this.checkStatements(node.statements);
        super.visitCaseClause(node);
    }
    checkStatements(statements) {
        const usedKeys = new Map();
        let collection = null;
        statements.forEach(statement => {
            const keyWriteUsage = this.keyWriteUsage(statement);
            if (keyWriteUsage) {
                if (collection && !areEquivalent_1.default(keyWriteUsage.collectionNode, collection)) {
                    usedKeys.clear();
                }
                const sameKeyWriteUsage = usedKeys.get(keyWriteUsage.indexOrKey);
                if (sameKeyWriteUsage) {
                    this.addIssue(keyWriteUsage.node, Rule.message(keyWriteUsage.indexOrKey, sameKeyWriteUsage.node)).addSecondaryLocation(sonarAnalysis_1.getIssueLocationAtNode(sameKeyWriteUsage.node, `Previous write to "${keyWriteUsage.indexOrKey}".`));
                }
                usedKeys.set(keyWriteUsage.indexOrKey, keyWriteUsage);
                collection = keyWriteUsage.collectionNode;
            }
            else {
                usedKeys.clear();
            }
        });
    }
    keyWriteUsage(node) {
        if (nodes_1.isExpressionStatement(node)) {
            return (this.arrayKeyWriteUsage(node.expression) ||
                this.mapKeyWriteUsage(node.expression) ||
                this.setKeyWriteUsage(node.expression) ||
                this.objectKeyWriteUsage(node.expression));
        }
    }
    arrayKeyWriteUsage(node) {
        if (nodes_1.isAssignment(node) && nodes_1.isElementAccessExpression(node.left)) {
            const lhs = node.left;
            const array = this.program.getTypeChecker().getSymbolAtLocation(lhs.expression);
            if (!array || this.usedInRhs(node.right, array))
                return;
            const index = this.extractIndex(lhs.argumentExpression);
            if (!index)
                return;
            return {
                collectionNode: lhs.expression,
                indexOrKey: index,
                node: lhs.expression,
            };
        }
    }
    usedInRhs(rhs, symbol) {
        return navigation_1.descendants(rhs)
            .filter(child => child.kind === ts.SyntaxKind.Identifier)
            .some(id => this.program.getTypeChecker().getSymbolAtLocation(id) === symbol);
    }
    mapKeyWriteUsage(node) {
        return this.callExpression(node, "Map", "set");
    }
    setKeyWriteUsage(node) {
        return this.callExpression(node, "Set", "add");
    }
    callExpression(node, typeName, method) {
        if (nodes_1.isCallExpression(node) && nodes_1.isPropertyAccessExpression(node.expression)) {
            const propertyAccess = node.expression;
            const type = this.program.getTypeChecker().getTypeAtLocation(propertyAccess.expression);
            if (type.symbol && type.symbol.name === typeName && propertyAccess.name.text === method) {
                const key = this.extractIndex(node.arguments[0]);
                if (!key)
                    return;
                return {
                    collectionNode: propertyAccess.expression,
                    indexOrKey: key,
                    node: propertyAccess.expression,
                };
            }
        }
    }
    objectKeyWriteUsage(node) {
        if (nodes_1.isAssignment(node) && nodes_1.isPropertyAccessExpression(node.left)) {
            const lhs = node.left;
            // avoid deeply nested property access
            if (!nodes_1.isIdentifier(lhs.expression))
                return;
            const objectSymbol = this.program.getTypeChecker().getSymbolAtLocation(lhs.expression);
            if (!objectSymbol)
                return;
            if (this.usedInRhs(node.right, objectSymbol))
                return;
            const property = lhs.name.text;
            if (!property)
                return;
            return {
                collectionNode: lhs.expression,
                indexOrKey: property,
                node: lhs.expression,
            };
        }
    }
    extractIndex(node) {
        if (!node)
            return;
        if (nodes_1.isNumericLiteral(node) || nodes_1.isStringLiteral(node)) {
            return node.text;
        }
        const symbol = this.program.getTypeChecker().getSymbolAtLocation(node);
        return symbol && symbol.name;
    }
}
