// tslint-sonarts/lib/rules/noGratuitousExpressionsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const SymbolicExecution_1 = require("../se/SymbolicExecution");
const builder_1 = require("../cfg/builder");
const programStates_1 = require("../se/programStates");
const constraints_1 = require("../se/constraints");
const builder_2 = require("../symbols/builder");
const table_1 = require("../symbols/table");
const navigation_1 = require("../utils/navigation");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    static getMessage(value) {
        return `This condition always evaluates to "${value}".`;
    }
    applyWithProgram(sourceFile, program) {
        const symbols = builder_2.SymbolTableBuilder.build(sourceFile, program);
        return new Visitor(this.getOptions(), program, symbols).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-gratuitous-expressions",
    description: 'Conditions should not always evaluate to "true" or to "false" ',
    rationale: tslint.Utils.dedent `
      If an expression doesn't change the evaluation of the condition,
      then it is either unnecessary, and condition can be removed,
      or it makes some code being never executed. In any case, the code should be refactored.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-2589",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    constructor(options, program, symbols) {
        super(options.ruleName, program);
        this.symbols = symbols;
    }
    visitFunctionLikeDeclaration(node) {
        const statements = this.getStatements(node);
        if (statements) {
            const initialState = programStates_1.createInitialState(node, this.program);
            const shouldTrackSymbol = (symbol) => this.symbols
                .allUsages(symbol)
                .filter(usage => usage.is(table_1.UsageFlag.WRITE))
                .every(usage => navigation_1.firstLocalAncestor(usage.node, ...navigation_1.FUNCTION_LIKE) === node);
            this.runForStatements(Array.from(statements), initialState, shouldTrackSymbol);
        }
        super.visitFunctionLikeDeclaration(node);
    }
    getStatements(functionLike) {
        if (nodes_1.isArrowFunction(functionLike)) {
            // `body` can be a block or an expression
            if (nodes_1.isBlock(functionLike.body)) {
                return functionLike.body.statements;
            }
        }
        else {
            return functionLike.body && functionLike.body.statements;
        }
        return undefined;
    }
    runForStatements(statements, initialState, shouldTrackSymbol) {
        const cfg = builder_1.build(statements);
        if (cfg) {
            const result = SymbolicExecution_1.execute(cfg, this.symbols, initialState, shouldTrackSymbol);
            if (result) {
                result.branchingProgramNodes.forEach((states, branchingProgramPoint) => {
                    if (this.ifAllProgramStateConstraints(states, constraints_1.isTruthy)) {
                        this.addIssue(branchingProgramPoint, Rule.getMessage("true"));
                    }
                    else if (this.ifAllProgramStateConstraints(states, constraints_1.isFalsy)) {
                        this.addIssue(branchingProgramPoint, Rule.getMessage("false"));
                    }
                });
            }
        }
    }
    ifAllProgramStateConstraints(programStates, checker) {
        return programStates.every(programState => {
            const [sv] = programState.popSV();
            return sv !== undefined && checker(programState.getConstraints(sv));
        });
    }
}
