// tslint-sonarts/lib/rules/noHardcodedCredentialsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
const DEFAULT_VARIABLE_NAMES = ["password", "pwd", "passwd"];
class Rule extends tslint.Rules.AbstractRule {
    get variableNames() {
        return this.ruleArguments.length > 0 ? this.ruleArguments : DEFAULT_VARIABLE_NAMES;
    }
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName, this.variableNames).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-hardcoded-credentials",
    description: "Credentials should not be hard-coded",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "Words matching variables potentially storing hardcoded credentials",
    options: {
        type: "array",
        items: { type: "string" },
    },
    optionExamples: [true, [true, ...DEFAULT_VARIABLE_NAMES]],
    rspecKey: "RSPEC-2068",
    type: "functionality",
    typescriptOnly: false,
};
Rule.MESSAGE = "Review this potentially hardcoded credential.";
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor(ruleName, variableNames) {
        super(ruleName);
        this.variableNames = variableNames;
        this.literalRegExp = this.variableNames.map(name => new RegExp(name + "=.+"));
    }
    visitVariableDeclaration(node) {
        this.checkAssignment(node.name, node.initializer);
        super.visitVariableDeclaration(node);
    }
    visitPropertyAssignment(node) {
        this.checkAssignment(node.name, node.initializer);
        super.visitPropertyAssignment(node);
    }
    visitBinaryExpression(node) {
        if (node.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
            this.checkAssignment(node.left, node.right);
        }
        super.visitBinaryExpression(node);
    }
    visitStringLiteral(node) {
        const text = node.getText();
        if (this.literalRegExp.some(regex => regex.test(text))) {
            this.addIssue(node, Rule.MESSAGE);
        }
        super.visitStringLiteral(node);
    }
    checkAssignment(variable, value) {
        if (value &&
            value.getText().length > 2 &&
            nodes_1.isStringLiteral(value) &&
            this.variableNames.some(name => variable.getText().includes(name))) {
            this.addIssue(value, Rule.MESSAGE);
        }
    }
}
