// tslint-sonarts/lib/rules/noIdenticalFunctionsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const areEquivalent_1 = require("../utils/areEquivalent");
const navigation_1 = require("../utils/navigation");
const nodes_1 = require("../utils/nodes");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        const visitor = new Visitor(this.getOptions().ruleName);
        visitor.visit(sourceFile);
        const functionBlocks = visitor.functionBlocks;
        if (functionBlocks.length < 2) {
            return [];
        }
        for (let i = 1; i < functionBlocks.length; i++) {
            const duplicatingFunctionBlock = functionBlocks[i];
            for (let j = 0; j < i; j++) {
                const originalFunctionBlock = functionBlocks[j];
                if (areEquivalent_1.default(duplicatingFunctionBlock, originalFunctionBlock, true)) {
                    visitor
                        .addIssue(Rule.issueNode(duplicatingFunctionBlock.parent), Rule.message(originalFunctionBlock))
                        .addSecondaryLocation(sonarAnalysis_1.getIssueLocationAtNode(originalFunctionBlock.parent, "original implementation"));
                    break;
                }
            }
        }
        return visitor.getIssues();
    }
    static message(functionBlock) {
        const lineOfOriginalFunction = navigation_1.startLineAndCharacter(functionBlock.parent).line + 1;
        return `Update or refactor this function so that its implementation doesn't duplicate the one on line ${lineOfOriginalFunction}.`;
    }
    static issueNode(functionNode) {
        if (nodes_1.is(functionNode, ts.SyntaxKind.FunctionExpression, ts.SyntaxKind.FunctionDeclaration)) {
            return navigation_1.findChild(functionNode, ts.SyntaxKind.FunctionKeyword);
        }
        if (nodes_1.is(functionNode, ts.SyntaxKind.MethodDeclaration, ts.SyntaxKind.GetAccessor, ts.SyntaxKind.SetAccessor)) {
            return navigation_1.findChild(functionNode, ts.SyntaxKind.Identifier);
        }
        if (nodes_1.is(functionNode, ts.SyntaxKind.Constructor)) {
            return navigation_1.findChild(functionNode, ts.SyntaxKind.ConstructorKeyword);
        }
        if (nodes_1.isArrowFunction(functionNode)) {
            return functionNode.equalsGreaterThanToken;
        }
        throw new Error("Unknow function kind " + ts.SyntaxKind[functionNode.kind]);
    }
}
Rule.metadata = {
    ruleName: "no-identical-functions",
    description: "",
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4144",
    type: "maintainability",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor() {
        super(...arguments);
        this.functionBlocks = [];
    }
    visitFunctionLikeDeclaration(node) {
        const { body } = node;
        if (body && nodes_1.isBlock(body) && Visitor.isBigEnough(body)) {
            this.functionBlocks.push(body);
        }
        super.visitFunctionLikeDeclaration(node);
    }
    static isBigEnough(block) {
        if (block.statements.length > 0) {
            const firstLine = navigation_1.startLineAndCharacter(block.statements[0]).line;
            const lastLine = navigation_1.endLineAndCharacter(block.statements[block.statements.length - 1]).line;
            return lastLine - firstLine > 1;
        }
        return false;
    }
}
