// tslint-sonarts/lib/rules/noIgnoredInitialValueRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const builder_1 = require("../symbols/builder");
const table_1 = require("../symbols/table");
const navigation_1 = require("../utils/navigation");
const lva_1 = require("../symbols/lva");
const cfg_1 = require("../cfg/cfg");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    static formatMessage(deadIdentifier) {
        return `Introduce a new variable or use its initial value before reassigning "${deadIdentifier.getText()}".`;
    }
    applyWithProgram(sourceFile, program) {
        const symbols = builder_1.SymbolTableBuilder.build(sourceFile, program);
        return new Visitor(this.getOptions(), program, symbols).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-ignored-initial-value",
    description: "Function parameters, caught exceptions and foreach variables' initial values should not be ignored",
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-1226",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    constructor(options, program, symbols) {
        super(options.ruleName, program);
        this.symbols = symbols;
        this.lva = new lva_1.LiveVariableAnalyzer(this.symbols);
    }
    visitFunctionLikeDeclaration(node) {
        if (node.body) {
            const lvaReturn = this.lva.analyzeFunction(node);
            this.check(node.body, lvaReturn, ...node.parameters);
        }
        super.visitFunctionLikeDeclaration(node);
    }
    visitCatchClause(node) {
        if (node.variableDeclaration) {
            const cfg = cfg_1.ControlFlowGraph.fromStatements(Array.from(node.block.statements));
            if (cfg) {
                const lvaReturn = this.lva.analyze(node.block, cfg);
                this.check(node.block, lvaReturn, node.variableDeclaration);
            }
        }
        super.visitCatchClause(node);
    }
    visitForInStatement(node) {
        this.visitForInOfStatement(node);
        super.visitForInStatement(node);
    }
    visitForOfStatement(node) {
        this.visitForInOfStatement(node);
        super.visitForOfStatement(node);
    }
    visitForInOfStatement(node) {
        const cfg = cfg_1.ControlFlowGraph.fromStatements([node.statement]);
        if (cfg) {
            const lvaReturn = this.lva.analyze(node, cfg);
            this.check(node.statement, lvaReturn, node.initializer);
        }
    }
    check(root, lvaReturn, ...nodesToCheck) {
        if (!lvaReturn) {
            return;
        }
        const { cfg, blockAvailableReads } = lvaReturn;
        const symbolsLiveAtStart = blockAvailableReads.get(cfg.start);
        nodesToCheck.forEach(parameter => {
            navigation_1.descendants(parameter)
                .filter(nodes_1.isIdentifier)
                .forEach(identifier => {
                const symbol = this.program.getTypeChecker().getSymbolAtLocation(identifier);
                if (symbol &&
                    !symbolsLiveAtStart.has(symbol) &&
                    isVariableOrParameter(symbol) &&
                    this.onlyUsedLocallyToRoot(symbol, root) &&
                    this.symbols.allUsages(symbol).length > 1) {
                    this.addIssue(identifier, Rule.formatMessage(identifier));
                }
            });
        });
    }
    onlyUsedLocallyToRoot(symbol, root) {
        const boundaries = [
            ...navigation_1.FUNCTION_LIKE,
            ts.SyntaxKind.SourceFile,
            ts.SyntaxKind.ClassDeclaration,
            ts.SyntaxKind.ClassExpression,
        ];
        return this.symbols
            .allUsages(symbol)
            .every(usage => usage.is(table_1.UsageFlag.DECLARATION) || navigation_1.ancestorsChain(usage.node, ...boundaries).includes(root));
    }
}
function isVariableOrParameter(symbol) {
    return symbol.flags === ts.SymbolFlags.BlockScopedVariable || symbol.flags === ts.SymbolFlags.FunctionScopedVariable;
}
