// tslint-sonarts/lib/rules/noInconsistentReturnRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const cfg_1 = require("../cfg/cfg");
const navigation_1 = require("../utils/navigation");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    description: "Functions should use 'return' consistently",
    options: null,
    optionsDescription: "",
    rspecKey: "RSPEC-3801",
    ruleName: "no-inconsistent-return",
    type: "maintainability",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitFunctionDeclaration(func) {
        if (!func.body)
            return;
        this.checkFunctionLikeDeclaration(func, func.body, func.type);
    }
    visitMethodDeclaration(meth) {
        if (!meth.body)
            return;
        this.checkFunctionLikeDeclaration(meth, meth.body, meth.type);
    }
    visitGetAccessor(accessor) {
        if (!accessor.body)
            return;
        this.checkFunctionLikeDeclaration(accessor, accessor.body, accessor.type);
    }
    visitArrowFunction(func) {
        if (func.body.kind === ts.SyntaxKind.Block) {
            this.checkFunctionLikeDeclaration(func, func.body, func.type);
        }
    }
    checkFunctionLikeDeclaration(functionNode, body, returnType) {
        if (this.declaredReturnTypeContainsVoidTypes(returnType))
            return;
        const cfg = cfg_1.ControlFlowGraph.fromStatements(Array.from(body.statements));
        if (cfg) {
            const predecessors = cfg.end.predecessors.filter(block => block === cfg.start || this.blockHasPredecessors(block));
            const hasExplicit = predecessors.find(this.lastElementIsExplicitReturn);
            const hasImplicit = predecessors.find(this.lastElementIsNotExplicitReturn);
            if (hasExplicit && hasImplicit) {
                this.addIssue(navigation_1.functionLikeMainToken(functionNode), 'Refactor this function to use "return" consistently');
            }
        }
    }
    declaredReturnTypeContainsVoidTypes(returnType) {
        if (returnType) {
            if (returnType.kind === ts.SyntaxKind.UnionType && returnType.types.find(isVoidType)) {
                return true;
            }
            else if (isVoidType(returnType)) {
                return true;
            }
        }
        return false;
        function isVoidType(type) {
            return type.kind === ts.SyntaxKind.UndefinedKeyword || type.kind === ts.SyntaxKind.VoidKeyword;
        }
    }
    lastElementIsNotExplicitReturn(cfgBlock) {
        const elements = cfgBlock.getElements();
        const lastElement = elements[elements.length - 1];
        if (!lastElement) {
            return false;
        }
        if (lastElement.kind === ts.SyntaxKind.ThrowStatement) {
            return false;
        }
        return lastElement.kind !== ts.SyntaxKind.ReturnStatement || !lastElement.expression;
    }
    lastElementIsExplicitReturn(cfgBlock) {
        const elements = cfgBlock.getElements();
        const lastElement = elements[elements.length - 1];
        if (!lastElement) {
            return false;
        }
        if (lastElement.kind === ts.SyntaxKind.ThrowStatement) {
            return false;
        }
        return lastElement.kind === ts.SyntaxKind.ReturnStatement && !!lastElement.expression;
    }
    blockHasPredecessors(cfgBlock) {
        if (cfgBlock.predecessors) {
            return cfgBlock.predecessors.length > 0;
        }
        return false;
    }
}
