// tslint-sonarts/lib/rules/noMisleadingArrayReverseRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
const semantics_1 = require("../utils/semantics");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    static getMessage(methodName) {
        return `Move this array "${methodName}" operation to a separate statement.`;
    }
    applyWithProgram(sourceFile, program) {
        return new Visitor(this.getOptions().ruleName, program).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-misleading-array-reverse",
    description: "Array-mutating methods should not be used misleadingly",
    rationale: tslint.Utils.dedent `
      Many of JavaScript's Array methods return an altered version of the array while leaving the source array intact.
      reverse and sort are not one of these. Instead, they alter the source array in addition to returning the altered version, which is likely not what was intended.
      To make sure maintainers are explicitly aware of this change to the original array, calls to reverse() should be
      standalone statements or preceded by a call that duplicates the original array.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4043",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    visitCallExpression(callExpression) {
        // first, take all call expressions: `foo()`
        // then check that:
        // * callee is a property access expression
        // * left part of callee is array
        // * the property is mutating, e.g."reverse" or "sort": `foo.reverse()`
        if (nodes_1.isPropertyAccessExpression(callExpression.expression) && this.isArrayMutatingCall(callExpression.expression)) {
            // store `foo` from `foo.reverse()`, `foo.sort()`, or `foo.bar` from `foo.bar.reverse()`, etc
            const mutatedArray = callExpression.expression.expression;
            if (
            // check that the left part of the property access expression is:
            // * identifier: `foo.reverse()`
            // * another property access expression: `foo.bar.reverse()`
            this.isIdentifierOrPropertyAccessExpression(mutatedArray) &&
                // exlude case `a = a.reverse()`
                !this.isReverseInSelfAssignment(mutatedArray, callExpression.parent) &&
                // check if we face one of the forbidden usages
                this.isForbiddenOperation(callExpression)) {
                this.addIssue(callExpression, Rule.getMessage(callExpression.expression.name.text));
            }
        }
        super.visitCallExpression(callExpression);
    }
    isArrayMutatingCall(expression) {
        return (semantics_1.isArray(expression.expression, this.program.getTypeChecker()) &&
            semantics_1.ARRAY_MUTATING_CALLS.includes(expression.name.text));
    }
    isGetAccessor(node) {
        const symbol = this.program.getTypeChecker().getSymbolAtLocation(node);
        const declarations = symbol && symbol.declarations;
        return (declarations !== undefined && declarations.length === 1 && declarations[0].kind === ts.SyntaxKind.GetAccessor);
    }
    isIdentifierOrPropertyAccessExpression(node) {
        // exclude class getters from consideration
        return nodes_1.isIdentifier(node) || (nodes_1.isPropertyAccessExpression(node) && !this.isGetAccessor(node));
    }
    isForbiddenOperation(node) {
        const { parent } = node;
        return (parent != null &&
            parent.kind !== ts.SyntaxKind.ExpressionStatement &&
            !navigation_1.firstLocalAncestor(node, ts.SyntaxKind.ReturnStatement));
    }
    isReverseInSelfAssignment(reversedArray, node) {
        return (
        // check assignment
        node !== undefined &&
            nodes_1.isBinaryExpression(node) &&
            node.operatorToken.kind === ts.SyntaxKind.EqualsToken &&
            // check that identifiers on both sides are the same
            nodes_1.isIdentifier(node.left) &&
            nodes_1.isIdentifier(reversedArray) &&
            node.left.text === reversedArray.text);
    }
}
