// tslint-sonarts/lib/rules/noMisspelledOperatorRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    static formatMessage(suggestedOperator) {
        return `Was "${suggestedOperator}" meant instead?`;
    }
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-misspelled-operator",
    description: "Non-existent operators '=+', '=-' and '=!' should not be used",
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-2757",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitBinaryExpression(node) {
        if (node.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
            const expression = node.right;
            if (expression.kind === ts.SyntaxKind.PrefixUnaryExpression) {
                const unaryExpression = expression;
                const unaryOperator = unaryExpression.getFirstToken();
                if (this.isPresentInCompoundAssignments(unaryOperator) &&
                    this.areAdjacent(node.operatorToken, unaryOperator) &&
                    !this.areAdjacent(unaryOperator, unaryExpression.operand)) {
                    this.addOperatorFailure(node.operatorToken, unaryOperator);
                }
            }
        }
        super.visitBinaryExpression(node);
    }
    isPresentInCompoundAssignments(operator) {
        return Visitor.COMPOUND_ASSIGNMENT_OPERATORS.has(operator.kind);
    }
    areAdjacent(left, right) {
        return left.getEnd() === right.getStart();
    }
    suggestOperator(unaryOperator) {
        return Visitor.COMPOUND_ASSIGNMENT_OPERATORS.get(unaryOperator.kind);
    }
    addOperatorFailure(previousOperator, unaryOperator) {
        this.addIssueAtLocation(new sonarAnalysis_1.IssueLocation(previousOperator.getStart(), unaryOperator.getEnd(), previousOperator.getSourceFile(), Rule.formatMessage(this.suggestOperator(unaryOperator))));
    }
}
Visitor.COMPOUND_ASSIGNMENT_OPERATORS = new Map()
    .set(ts.SyntaxKind.PlusToken, "+=")
    .set(ts.SyntaxKind.MinusToken, "-=")
    .set(ts.SyntaxKind.ExclamationToken, "!=");
