// tslint-sonarts/lib/rules/noMultilineStringLiteralsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-multiline-string-literals",
    description: "Multiline string literals should not be used",
    rationale: tslint.Utils.dedent `
      Continuing a string across a linebreak is supported in most script engines, but it is not a part of ECMAScript.
      Additionally, the whitespace at the beginning of each line can't be safely stripped at compile time, and any
      whitespace after the slash will result in tricky errors.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-1516",
    type: "functionality",
    typescriptOnly: false,
};
Rule.MESSAGE = "Use string concatenation rather than line continuation.";
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitStringLiteral(node) {
        if (this.isMultiline(node)) {
            this.addIssue(node, Rule.MESSAGE);
        }
        super.visitStringLiteral(node);
    }
    isMultiline(node) {
        const startLine = node.getSourceFile().getLineAndCharacterOfPosition(node.getStart()).line;
        const endLine = node.getSourceFile().getLineAndCharacterOfPosition(node.getEnd()).line;
        return endLine > startLine;
    }
}
