// tslint-sonarts/lib/rules/noNestedIncdecRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-nested-incdec",
    description: "Increment (++) and decrement (--) operators should not be used in a method call or mixed with other operators in an expression",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-881",
    type: "functionality",
    typescriptOnly: false,
};
Rule.MESSAGE_INC = "Extract this increment operation into a dedicated statement.";
Rule.MESSAGE_DEC = "Extract this decrement operation into a dedicated statement.";
exports.Rule = Rule;
const incDecOperators = [ts.SyntaxKind.PlusPlusToken, ts.SyntaxKind.MinusMinusToken];
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitExpressionStatement(node) {
        if (isIncDec(node.expression)) {
            // skipping visit visitPrefixUnaryExpression/visitPostfixUnaryExpression if increment/decrement
            this.visitNode(node.expression.operand);
        }
        else {
            super.visitExpressionStatement(node);
        }
    }
    visitForStatement(node) {
        node.initializer && this.visitNode(node.initializer);
        node.condition && this.visitNode(node.condition);
        node.incrementor && this.visitIncrementor(node.incrementor);
        this.visitNode(node.statement);
    }
    visitPrefixUnaryExpression(node) {
        this.checkUnaryExpression(node);
        super.visitPrefixUnaryExpression(node);
    }
    visitPostfixUnaryExpression(node) {
        this.checkUnaryExpression(node);
        super.visitPostfixUnaryExpression(node);
    }
    visitIncrementor(node) {
        if (isIncDec(node)) {
            this.visitNode(node.operand);
        }
        else if (nodes_1.isBinaryExpression(node) && node.operatorToken.kind === ts.SyntaxKind.CommaToken) {
            this.visitIncrementor(node.left);
            this.visitIncrementor(node.right);
        }
        else {
            this.visitNode(node);
        }
    }
    checkUnaryExpression(node) {
        if (node.operator === ts.SyntaxKind.PlusPlusToken) {
            this.addIssue(node, Rule.MESSAGE_INC);
        }
        else if (node.operator === ts.SyntaxKind.MinusMinusToken) {
            this.addIssue(node, Rule.MESSAGE_DEC);
        }
    }
}
function isIncDec(node) {
    return (nodes_1.isPrefixUnaryExpression(node) || nodes_1.isPostfixUnaryExpression(node)) && incDecOperators.includes(node.operator);
}
