// tslint-sonarts/lib/rules/noRedundantJumpRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const cfg_1 = require("../cfg/cfg");
const nodes_1 = require("../utils/nodes");
const navigation_1 = require("../utils/navigation");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-redundant-jump",
    description: "Jump statements should not be redundant",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-3626",
    type: "functionality",
    typescriptOnly: false,
};
Rule.MESSAGE = "Remove this redundant jump.";
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitFunctionDeclaration(func) {
        if (!func.body)
            return;
        const cfg = cfg_1.ControlFlowGraph.fromStatements(Array.from(func.body.statements));
        if (cfg) {
            for (const block of cfg.getBlocks()) {
                const successors = block.getSuccessors();
                const lastElement = block.getElements()[block.getElements().length - 1];
                if (successors.length === 1 &&
                    successors[0] === block.successorWithoutJump &&
                    !isReturnWithExpression(lastElement) &&
                    !isInSwitch(lastElement) &&
                    !isContinueWithLabel(lastElement)) {
                    this.addIssue(lastElement, Rule.MESSAGE);
                }
            }
        }
    }
}
function isReturnWithExpression(node) {
    return nodes_1.isReturnStatement(node) && !!node.expression;
}
function isInSwitch(node) {
    const LOOPS = [
        ts.SyntaxKind.ForStatement,
        ts.SyntaxKind.WhileStatement,
        ts.SyntaxKind.DoStatement,
        ts.SyntaxKind.ForOfStatement,
        ts.SyntaxKind.ForInStatement,
    ];
    const localAncestor = navigation_1.firstLocalAncestor(node, ts.SyntaxKind.SwitchStatement, ...LOOPS);
    return nodes_1.is(localAncestor, ts.SyntaxKind.SwitchStatement);
}
function isContinueWithLabel(node) {
    return nodes_1.isContinueStatement(node) && !!node.label;
}
