// tslint-sonarts/lib/rules/noRedundantParenthesesRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const Lint = require("tslint");
const nodes_1 = require("../utils/nodes");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends Lint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    description: "Redundant pairs of parentheses should be removed",
    options: null,
    optionsDescription: "",
    rspecKey: "RSPEC-1110",
    ruleName: "no-redundant-parentheses",
    type: "maintainability",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitParenthesizedExpression(node) {
        if (nodes_1.isParenthesizedExpression(node.expression)) {
            this.addIssue(node, Visitor.MESSAGE);
        }
        super.visitParenthesizedExpression(node);
    }
}
Visitor.MESSAGE = "Remove these useless parentheses.";
