// tslint-sonarts/lib/rules/noSameLineConditionalRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const Lint = require("tslint");
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends Lint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    description: "Conditionals should start on new lines",
    options: null,
    optionsDescription: "",
    rspecKey: "RSPEC-3972",
    ruleName: "no-same-line-conditional",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitSourceFile(node) {
        this.checkStatements(Array.from(node.statements));
        super.visitSourceFile(node);
    }
    visitBlock(node) {
        this.checkStatements(Array.from(node.statements));
        super.visitBlock(node);
    }
    visitModuleDeclaration(node) {
        if (node.body && node.body.kind === ts.SyntaxKind.ModuleBlock) {
            this.checkStatements(Array.from(node.body.statements));
        }
        super.visitModuleDeclaration(node);
    }
    checkStatements(statements) {
        statements.forEach((statement, index) => {
            if (nodes_1.isIfStatement(statement) && index > 0) {
                const previousStatement = statements[index - 1];
                if (nodes_1.isIfStatement(previousStatement)) {
                    const ifTokenLine = navigation_1.startLineAndCharacter(statement).line;
                    const previousStatementLastLine = navigation_1.endLineAndCharacter(previousStatement).line;
                    if (ifTokenLine === previousStatementLastLine) {
                        this.addIssue(statement.getFirstToken(), Visitor.MESSAGE);
                    }
                }
            }
        });
    }
}
Visitor.MESSAGE = 'Move this "if" to a new line or add the missing "else".';
