// tslint-sonarts/lib/rules/noSelfAssignmentRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const areEquivalent_1 = require("../utils/areEquivalent");
const semantics_1 = require("../utils/semantics");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    static formatMessage() {
        return "Remove or correct this useless self-assignment.";
    }
    applyWithProgram(sourceFile, program) {
        return new Visitor(this.getOptions().ruleName, program).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-self-assignment",
    description: "Variables should not be self-assigned",
    rationale: tslint.Utils.dedent `
      There is no reason to re-assign a variable to itself. Either this statement is redundant and
      should be removed, or the re-assignment is a mistake and some other value or variable was
      intended for the assignment instead.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-1656",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    visitBinaryExpression(expression) {
        if (this.isAssignment(expression) && !this.hasAccessors(expression.left) && this.isSelfAssignment(expression)) {
            this.addIssue(expression, Rule.formatMessage());
        }
        super.visitBinaryExpression(expression);
    }
    isSelfAssignment(expression) {
        return (areEquivalent_1.default(expression.left, expression.right) ||
            this.isArrayWithSpreadExpressionOnly(expression.right, expression.left) ||
            this.isArrayReverseAssignment(expression.left, expression.right));
    }
    isAssignment(expression) {
        return nodes_1.is(expression.operatorToken, ts.SyntaxKind.EqualsToken);
    }
    hasAccessors(node) {
        const symbol = this.program.getTypeChecker().getSymbolAtLocation(node);
        const declarations = symbol && symbol.declarations;
        return (declarations &&
            declarations.some(declaration => nodes_1.is(declaration, ts.SyntaxKind.GetAccessor, ts.SyntaxKind.SetAccessor)));
    }
    isArrayWithSpreadExpressionOnly(expression, variable) {
        if (nodes_1.isArrayLiteralExpression(expression)) {
            const { elements } = expression;
            if (elements.length === 1 && nodes_1.isSpreadElement(elements[0])) {
                return areEquivalent_1.default(elements[0].expression, variable);
            }
        }
        return false;
    }
    isArrayReverseAssignment(left, right) {
        // in case of `a = a.reverse()`, left is `a` and right is `a.reverse()`
        return (nodes_1.isCallExpression(right) &&
            nodes_1.isPropertyAccessExpression(right.expression) &&
            nodes_1.isIdentifier(right.expression.expression) &&
            this.isArrayMutatingCall(right.expression) &&
            areEquivalent_1.default(right.expression.expression, left));
    }
    isArrayMutatingCall(expression) {
        return (semantics_1.isArray(expression.expression, this.program.getTypeChecker()) &&
            semantics_1.ARRAY_MUTATING_CALLS.includes(expression.name.text));
    }
}
