// tslint-sonarts/lib/rules/noStatementsSameLineRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
const navigation_1 = require("../utils/navigation");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        const visitor = new Visitor(this.getOptions().ruleName).visit(sourceFile);
        visitor.statementsPerLine.forEach(statements => {
            if (statements.length > 1) {
                const issue = visitor.addIssue(statements[1], Rule.MESSAGE);
                statements.slice(2).forEach(statement => issue.addSecondaryLocation(statement));
            }
        });
        return visitor.getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-statements-same-line",
    description: "Statements should be on separate lines",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-122",
    type: "maintainability",
    typescriptOnly: false,
};
Rule.MESSAGE = "Reformat the code to have only one statement per line.";
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor() {
        super(...arguments);
        this.statementsPerLine = new Map();
        this.excludedStatements = new Set();
    }
    visitNode(node) {
        if (nodes_1.isIfStatement(node)) {
            this.checkForExcludedStatement(node.thenStatement, node);
        }
        if (nodes_1.isIterationStatement(node)) {
            this.checkForExcludedStatement(node.statement, node);
        }
        if (nodes_1.isFunctionExpression(node) || nodes_1.isArrowFunction(node)) {
            this.checkFunctionException(node);
        }
        if (nodes_1.is(node, ...Visitor.STATEMENT_KINDS) && !this.excludedStatements.has(node)) {
            const line = navigation_1.startLineAndCharacter(node).line;
            const statementsOnThisLine = this.statementsPerLine.get(line) || [];
            statementsOnThisLine.push(node);
            this.statementsPerLine.set(line, statementsOnThisLine);
        }
        super.visitNode(node);
    }
    checkFunctionException(functionExpression) {
        if (nodes_1.isBlock(functionExpression.body)) {
            const startLineOfFunction = navigation_1.startLineAndCharacter(functionExpression).line;
            const statements = functionExpression.body.statements;
            if (statements.length === 1 &&
                navigation_1.startLineAndCharacter(statements[0]).line === startLineOfFunction &&
                this.statementsPerLine.has(startLineOfFunction)) {
                this.excludedStatements.add(statements[0]);
            }
        }
    }
    checkForExcludedStatement(nestedStatement, nestingStatement) {
        const startOfNestingStatement = navigation_1.startLineAndCharacter(nestingStatement);
        const endOfNestedStatement = navigation_1.endLineAndCharacter(nestedStatement);
        if (nodes_1.isBlock(nestedStatement)) {
            if (endOfNestedStatement.line === startOfNestingStatement.line && nestedStatement.statements.length === 1) {
                this.excludedStatements.add(nestedStatement.statements[0]);
            }
        }
        else {
            const startOfNestedStatement = navigation_1.startLineAndCharacter(nestedStatement);
            if (startOfNestedStatement.line === startOfNestingStatement.line) {
                this.excludedStatements.add(nestedStatement);
            }
        }
    }
}
Visitor.STATEMENT_KINDS = [
    ts.SyntaxKind.VariableStatement,
    ts.SyntaxKind.ExpressionStatement,
    ts.SyntaxKind.IfStatement,
    ts.SyntaxKind.DoStatement,
    ts.SyntaxKind.WhileStatement,
    ts.SyntaxKind.ForStatement,
    ts.SyntaxKind.ForInStatement,
    ts.SyntaxKind.ForOfStatement,
    ts.SyntaxKind.ContinueStatement,
    ts.SyntaxKind.BreakStatement,
    ts.SyntaxKind.ReturnStatement,
    ts.SyntaxKind.WithStatement,
    ts.SyntaxKind.SwitchStatement,
    ts.SyntaxKind.ThrowStatement,
    ts.SyntaxKind.TryStatement,
    ts.SyntaxKind.DebuggerStatement,
];
