// tslint-sonarts/lib/rules/noUnconditionalJumpRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const cfg_1 = require("../cfg/cfg");
const nav = require("../utils/navigation");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    description: "Loops with at most one iteration should be refactored",
    options: null,
    optionsDescription: "",
    rspecKey: "RSPEC-1751",
    ruleName: "no-unconditional-jump",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    constructor() {
        super(...arguments);
        this.loopsAndJumps = new Map();
    }
    visitBreakStatement(node) {
        this.checkJump(node);
    }
    visitThrowStatement(node) {
        this.checkJump(node);
    }
    visitReturnStatement(node) {
        this.checkJump(node);
    }
    checkJump(node) {
        const keyword = nav.keyword(node);
        if (node.kind === ts.SyntaxKind.BreakStatement && this.isInsideForIn(node))
            return;
        if (node.kind === ts.SyntaxKind.ReturnStatement && this.isInsideForOf(node))
            return;
        const cfg = this.buildCfg(node);
        if (!cfg)
            return;
        const loop = nav.firstLocalAncestor(node, ...nav.LOOP_STATEMENTS);
        if (loop) {
            const loopingBlock = cfg.findLoopingBlock(loop);
            if (loopingBlock && this.actuallyLoops(loopingBlock))
                return;
            const jumpsForThisLoop = this.loopsAndJumps.get(loop) || [];
            jumpsForThisLoop.push(keyword);
            this.loopsAndJumps.set(loop, jumpsForThisLoop);
        }
    }
    visitSourceFile(node) {
        super.visitSourceFile(node);
        this.loopsAndJumps.forEach((jumps, loop) => {
            const issue = this.addIssue(loop.getFirstToken(), "Refactor this loop; it's executed only once");
            jumps.forEach(jump => issue.addSecondaryLocation(jump, "loop is broken here."));
        });
    }
    isInsideForIn(node) {
        const parentLoop = nav.firstLocalAncestor(node, ...nav.LOOP_STATEMENTS);
        return !!parentLoop && parentLoop.kind === ts.SyntaxKind.ForInStatement;
    }
    isInsideForOf(node) {
        const parentLoop = nav.firstLocalAncestor(node, ...nav.LOOP_STATEMENTS);
        return !!parentLoop && parentLoop.kind === ts.SyntaxKind.ForOfStatement;
    }
    buildCfg(node) {
        const wrappingFunction = nav.firstLocalAncestor(node, ...nav.FUNCTION_LIKE);
        if (wrappingFunction && wrappingFunction.body) {
            if (wrappingFunction.body.kind === ts.SyntaxKind.Block) {
                return cfg_1.ControlFlowGraph.fromStatements(Array.from(wrappingFunction.body.statements));
            }
            else {
                return; // When moving buildCfg to cfg file this should be replaced by fromExpression, here instead, we skip
            }
        }
        return cfg_1.ControlFlowGraph.fromStatements(Array.from(node.getSourceFile().statements));
    }
    actuallyLoops(block) {
        if (!block.loopingStatement)
            return false;
        const loopContents = this.collectLoopContents(block.loopingStatement);
        if (block instanceof cfg_1.CfgBlockWithPredecessors) {
            return !!block.predecessors.find(predecessor => this.hasPredecessor(predecessor) &&
                !!predecessor.getElements().find(predecessorElement => loopContents.includes(predecessorElement)));
        }
        else {
            return false;
        }
    }
    collectLoopContents(iterationStatement) {
        const bodyContents = nav.descendants(iterationStatement.statement);
        if (iterationStatement.kind === ts.SyntaxKind.ForStatement) {
            const updateExpression = iterationStatement.incrementor;
            if (updateExpression)
                return bodyContents.concat(nav.descendants(updateExpression));
        }
        return bodyContents;
    }
    hasPredecessor(block) {
        if (block instanceof cfg_1.CfgBlockWithPredecessors) {
            return block.predecessors.length > 0;
        }
        else {
            return false;
        }
    }
}
