// tslint-sonarts/lib/rules/noUndefinedArgumentRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    applyWithProgram(sourceFile, program) {
        return new Visitor(this.getOptions().ruleName, program).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-undefined-argument",
    description: `"undefined" should not be passed as the value of optional parameters`,
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4623",
    type: "maintainability",
    typescriptOnly: false,
};
Rule.MESSAGE = `Remove this redundant "undefined".`;
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    visitCallExpression(node) {
        const { arguments: args } = node;
        if (args.length === 0) {
            return;
        }
        const lastArgument = args[args.length - 1];
        if (nodes_1.isIdentifier(lastArgument) &&
            lastArgument.text === "undefined" &&
            this.isOptionalParameter(args.length - 1, node)) {
            this.addIssue(lastArgument, Rule.MESSAGE);
        }
        super.visitCallExpression(node);
    }
    isOptionalParameter(parameterIndex, node) {
        const declaration = this.program.getTypeChecker().getResolvedSignature(node).declaration;
        if (nodes_1.isFunctionLikeDeclaration(declaration)) {
            const { parameters } = declaration;
            const parameter = parameters[parameterIndex];
            return parameter && (parameter.initializer || parameter.questionToken);
        }
        return false;
    }
}
