// tslint-sonarts/lib/rules/noUnthrownErrorRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    description: "Errors should not be created without being thrown",
    options: null,
    optionsDescription: "",
    rspecKey: "RSPEC-3984",
    ruleName: "no-unthrown-error",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitNewExpression(node) {
        super.visitNewExpression(node);
        if (this.isWithinExpressionStatement(node) && this.looksLikeAnError(node.expression)) {
            this.addIssue(node, `Throw this error or remove this useless statement`);
        }
    }
    isWithinExpressionStatement(node) {
        return !!node.parent && node.parent.kind === ts.SyntaxKind.ExpressionStatement;
    }
    looksLikeAnError(expression) {
        const text = expression.getText();
        return text.endsWith("Error") || text.endsWith("Exception");
    }
}
