// tslint-sonarts/lib/rules/noUselessIncrementRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const areEquivalent_1 = require("../utils/areEquivalent");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-useless-increment",
    description: "Values should not be uselessly incremented",
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-2123",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitBinaryExpression(node) {
        if (node.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
            const rightOperand = node.right;
            const leftOperand = node.left;
            if (rightOperand.kind === ts.SyntaxKind.PostfixUnaryExpression) {
                const postfixUnaryExpr = rightOperand;
                if (areEquivalent_1.default(leftOperand, postfixUnaryExpr.operand)) {
                    this.addIssue(postfixUnaryExpr.getChildAt(1), postfixUnaryExpr.operator === ts.SyntaxKind.PlusPlusToken ? Visitor.MESSAGE_INC : Visitor.MESSAGE_DEC);
                }
            }
        }
    }
}
Visitor.MESSAGE_INC = "Remove this increment or correct the code not to waste it.";
Visitor.MESSAGE_DEC = "Remove this decrement or correct the code not to waste it.";
