// tslint-sonarts/lib/rules/noUselessIntersectionRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const semantics_1 = require("../utils/semantics");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    static formatAnyOrNeverMessage(type) {
        return `Simplify this intersection as it always has type "${type}".`;
    }
    applyWithProgram(sourceFile, program) {
        return new Visitor(this.getOptions().ruleName, program).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "no-useless-intersection",
    description: "Types without members, 'any' and 'never' should not be used in type intersections",
    rationale: tslint.Utils.dedent `
      An intersection type combines multiple types into one. This allows you to add together existing types to get a 
      single type that has all the features you need. However an intersection with a type without members doesn't 
      change the resulting type. In the opposite the usage of any or never as part of an intersection will always 
      results in any or never respectively. This is almost certainly an error.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4335",
    type: "functionality",
    typescriptOnly: false,
};
Rule.MESSAGE = "Remove this type without members or change this type intersection.";
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    visitIntersectionTypeNode(node) {
        const anyOrNever = node.types.find(typeNode => ["any", "never"].includes(typeNode.getText()));
        if (anyOrNever) {
            this.addIssue(node, Rule.formatAnyOrNeverMessage(anyOrNever.getText()));
        }
        else {
            node.types.forEach(typeNode => {
                const type = this.program.getTypeChecker().getTypeFromTypeNode(typeNode);
                if (isTypeWithoutMembers(type)) {
                    this.addIssue(typeNode, Rule.MESSAGE);
                }
            });
        }
        super.visitIntersectionTypeNode(node);
    }
}
function isTypeWithoutMembers(type) {
    const isNullLike = semantics_1.isNullType(type) || semantics_1.isUndefinedType(type) || semantics_1.isVoidType(type);
    const isEmptyInterface = Boolean(type.symbol && type.symbol.members && type.symbol.members.size === 0 && isStandaloneInterface(type.symbol));
    return isNullLike || isEmptyInterface;
}
function isStandaloneInterface({ declarations }) {
    // there is no declarations for `{}`
    // otherwise check that none of declarations has a heritage clause (`extends X` or `implments X`)
    return (!declarations ||
        declarations.every(declaration => {
            return nodes_1.isInterfaceDeclaration(declaration) && (declaration.heritageClauses || []).length === 0;
        }));
}
