// tslint-sonarts/lib/rules/preferImmediateReturnRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const builder_1 = require("../symbols/builder");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.TypedRule {
    applyWithProgram(sourceFile, program) {
        const symbols = builder_1.SymbolTableBuilder.build(sourceFile, program);
        return new Visitor(this.getOptions(), program, symbols).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "prefer-immediate-return",
    description: "Local variables should not be declared and then immediately returned or thrown",
    optionsDescription: "",
    options: null,
    rationale: tslint.Utils.dedent `
      Declaring a variable only to immediately return or throw it is a bad practice. Some developers argue that the 
      practice improves code readability, because it enables them to explicitly name what is being returned. However, 
      this variable is an internal implementation detail that is not exposed to the callers of the method. The method 
      name should be sufficient for callers to know exactly what will be returned.`,
    rspecKey: "RSPEC-1488",
    type: "functionality",
    typescriptOnly: false,
};
exports.Rule = Rule;
class Visitor extends sonarAnalysis_1.TypedSonarRuleVisitor {
    constructor(options, program, symbols) {
        super(options.ruleName, program);
        this.symbols = symbols;
    }
    visitBlock(node) {
        this.processStatements(node.statements);
        super.visitBlock(node);
    }
    visitCaseClause(node) {
        this.processStatements(node.statements);
        super.visitCaseClause(node);
    }
    visitDefaultClause(node) {
        this.processStatements(node.statements);
        super.visitDefaultClause(node);
    }
    processStatements(statements) {
        if (statements.length > 1) {
            const last = statements[statements.length - 1];
            const returnedVariable = this.getOnlyReturnedVariable(last);
            const lastButOne = statements[statements.length - 2];
            const declaredVariable = this.getOnlyDeclaredVariable(lastButOne);
            if (returnedVariable && declaredVariable) {
                const returnedSymbol = this.program.getTypeChecker().getSymbolAtLocation(returnedVariable);
                const declaredSymbol = this.program.getTypeChecker().getSymbolAtLocation(declaredVariable.name);
                if (returnedSymbol &&
                    returnedSymbol === declaredSymbol &&
                    !this.symbols.allUsages(declaredSymbol).some(usage => usage.isUsedInside(declaredVariable.initializer))) {
                    this.addIssue(declaredVariable.initializer, this.formatMessage(last, returnedVariable.text));
                }
            }
        }
    }
    getOnlyReturnedVariable(node) {
        return (nodes_1.isReturnStatement(node) || nodes_1.isThrowStatement(node)) && node.expression && nodes_1.isIdentifier(node.expression)
            ? node.expression
            : undefined;
    }
    getOnlyDeclaredVariable(node) {
        if (nodes_1.isVariableStatement(node) && node.declarationList.declarations.length === 1) {
            const { name, initializer } = node.declarationList.declarations[0];
            if (nodes_1.isIdentifier(name) && initializer) {
                return { name, initializer };
            }
        }
        return undefined;
    }
    formatMessage(node, variable) {
        const action = nodes_1.isReturnStatement(node) ? "return" : "throw";
        return `Immediately ${action} this expression instead of assigning it to the temporary variable "${variable}".`;
    }
}
