// tslint-sonarts/lib/rules/preferPromiseShorthandRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const sonarAnalysis_1 = require("../utils/sonarAnalysis");
const nodes_1 = require("../utils/nodes");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
    }
}
Rule.metadata = {
    ruleName: "prefer-promise-shorthand",
    description: "Shorthand promises should be used",
    rationale: tslint.Utils.dedent ``,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4634",
    type: "maintainability",
    typescriptOnly: true,
};
exports.Rule = Rule;
function message(value, action) {
    return `Replace this trivial promise with "Promise.${action}(${value.getText()})".`;
}
class Visitor extends sonarAnalysis_1.SonarRuleVisitor {
    visitNewExpression(node) {
        const executor = getPromiseExecutor(node);
        if (executor) {
            this.checkExecutor(executor, node);
        }
        super.visitNewExpression(node);
    }
    checkExecutor(executor, newExpression) {
        if (nodes_1.isFunctionLikeDeclaration(executor) && executor.body) {
            const { parameters, body } = executor;
            const [resolveParameterDeclaration, rejectParameterDeclaration] = parameters;
            const resolveParameterName = getParameterName(resolveParameterDeclaration);
            const rejectParameterName = getParameterName(rejectParameterDeclaration);
            const bodyExpression = getOnlyBodyExpression(body);
            if (bodyExpression && nodes_1.isCallExpression(bodyExpression)) {
                const { expression: callee, arguments: callArguments } = bodyExpression;
                if (nodes_1.isIdentifier(callee)) {
                    const action = getPromiseAction(callee.text, resolveParameterName, rejectParameterName);
                    if (action && callArguments.length === 1)
                        this.addIssue(newExpression, message(callArguments[0], action));
                }
            }
        }
    }
}
function getPromiseExecutor(node) {
    if (nodes_1.isIdentifier(node.expression) &&
        node.expression.text === "Promise" &&
        node.arguments &&
        node.arguments.length === 1) {
        return node.arguments[0];
    }
    return undefined;
}
function getOnlyBodyExpression(body) {
    let bodyExpression;
    if (nodes_1.isBlock(body)) {
        if (body.statements.length === 1) {
            const statement = body.statements[0];
            if (nodes_1.isExpressionStatement(statement)) {
                bodyExpression = statement.expression;
            }
        }
    }
    else {
        bodyExpression = body;
    }
    return bodyExpression;
}
function getPromiseAction(callee, resolveParameterName, rejectParameterName) {
    switch (callee) {
        case resolveParameterName:
            return "resolve";
        case rejectParameterName:
            return "reject";
    }
}
function getParameterName(parameterDeclaration) {
    return parameterDeclaration && nodes_1.isIdentifier(parameterDeclaration.name) ? parameterDeclaration.name.text : null;
}
