// tslint-sonarts/lib/runner/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const sonarUtils_1 = require("./sonarUtils");
const cyclomaticComplexity_1 = require("../utils/cyclomaticComplexity");
const cognitiveComplexity_1 = require("../utils/cognitiveComplexity");
const navigation_1 = require("../utils/navigation");
const nodes_1 = require("../utils/nodes");
function getMetrics(sourceFile) {
    return Object.assign({ ncloc: findLinesOfCode(sourceFile) }, findCommentLines(sourceFile), { executableLines: findExecutableLines(sourceFile), functions: countFunctions(sourceFile), statements: countStatements(sourceFile), classes: countClasses(sourceFile), complexity: cyclomaticComplexity_1.getOverallComplexity(sourceFile).length, cognitiveComplexity: cognitiveComplexity_1.getOverallCognitiveComplexity(sourceFile) });
}
exports.default = getMetrics;
function findLinesOfCode(sourceFile) {
    const lines = new Set();
    walk(sourceFile, node => {
        if (node.kind <= ts.SyntaxKind.OfKeyword && node.kind !== ts.SyntaxKind.EndOfFileToken) {
            addLines(node.getStart(), node.getEnd(), lines, sourceFile);
        }
    });
    return Array.from(lines).sort((a, b) => a - b);
}
exports.findLinesOfCode = findLinesOfCode;
function findCommentLines(sourceFile) {
    const commentLines = new Set();
    const nosonarLines = new Set();
    // ignore header comments -> comments before first token
    let first = true;
    walk(sourceFile, node => {
        if (node.kind <= ts.SyntaxKind.OfKeyword) {
            if (first) {
                first = false;
                navigation_1.getCommentsAfter(node).forEach(processComment);
            }
            else {
                navigation_1.getComments(node).forEach(processComment);
            }
        }
    });
    function processComment(comment) {
        const content = navigation_1.getText(comment, sourceFile).substr(2);
        if (content
            .trim()
            .toUpperCase()
            .startsWith("NOSONAR")) {
            addLines(comment.pos, comment.end, nosonarLines, sourceFile);
        }
        addLines(comment.pos, comment.end, commentLines, sourceFile);
    }
    return {
        commentLines: Array.from(commentLines).sort((a, b) => a - b),
        nosonarLines: Array.from(nosonarLines).sort((a, b) => a - b),
    };
}
exports.findCommentLines = findCommentLines;
function findExecutableLines(sourceFile) {
    const EXECUTABLE_STATEMENT_KINDS = [
        ts.SyntaxKind.DebuggerStatement,
        ts.SyntaxKind.VariableStatement,
        ts.SyntaxKind.LabeledStatement,
        ts.SyntaxKind.ReturnStatement,
        ts.SyntaxKind.BreakStatement,
        ts.SyntaxKind.ContinueStatement,
        ts.SyntaxKind.ThrowStatement,
        ts.SyntaxKind.WithStatement,
        ts.SyntaxKind.TryStatement,
        ts.SyntaxKind.SwitchStatement,
        ts.SyntaxKind.IfStatement,
        ts.SyntaxKind.WhileStatement,
        ts.SyntaxKind.DoStatement,
        ts.SyntaxKind.ExpressionStatement,
        ts.SyntaxKind.ForStatement,
        ts.SyntaxKind.ForInStatement,
        ts.SyntaxKind.ForOfStatement,
    ];
    const lines = new Set();
    walk(sourceFile, node => {
        if (EXECUTABLE_STATEMENT_KINDS.includes(node.kind)) {
            lines.add(sonarUtils_1.toSonarLine(navigation_1.startLineAndCharacter(node).line));
        }
    });
    return Array.from(lines).sort((a, b) => a - b);
}
exports.findExecutableLines = findExecutableLines;
function countClasses(sourceFile) {
    return walkAndCountIf(sourceFile, node => nodes_1.is(node, ts.SyntaxKind.ClassDeclaration, ts.SyntaxKind.ClassExpression));
}
exports.countClasses = countClasses;
function countFunctions(sourceFile) {
    const FUNCTION_KINDS = [
        ts.SyntaxKind.FunctionDeclaration,
        ts.SyntaxKind.FunctionExpression,
        ts.SyntaxKind.MethodDeclaration,
        ts.SyntaxKind.ArrowFunction,
    ];
    return walkAndCountIf(sourceFile, node => nodes_1.is(node, ...FUNCTION_KINDS));
}
exports.countFunctions = countFunctions;
function countStatements(sourceFile) {
    return walkAndCountIf(sourceFile, node => node.kind >= ts.SyntaxKind.VariableStatement && node.kind <= ts.SyntaxKind.DebuggerStatement);
}
exports.countStatements = countStatements;
function walk(node, walker) {
    const stack = [node];
    const toWalk = [];
    while (stack.length) {
        const currentNode = stack.pop();
        toWalk.push(currentNode);
        stack.push(...currentNode.getChildren());
    }
    toWalk.reverse();
    toWalk.forEach(walker);
}
function walkAndCountIf(root, condition) {
    let results = 0;
    walk(root, node => {
        if (condition(node)) {
            results++;
        }
    });
    return results;
}
function addLines(start, end, lines, sourceFile) {
    const firstLine = sonarUtils_1.toSonarLine(navigation_1.lineAndCharacterByPos(start, sourceFile).line);
    const lastLine = sonarUtils_1.toSonarLine(navigation_1.lineAndCharacterByPos(end, sourceFile).line);
    for (let i = firstLine; i <= lastLine; i++) {
        lines.add(i);
    }
}
