// tslint-sonarts/lib/runner/sonartsServer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const net = require("net");
const path = require("path");
const fs = require("fs");
const rules_1 = require("./rules");
const languageService_1 = require("./languageService");
const parser_1 = require("../utils/parser");
const diagnostics_1 = require("./diagnostics");
const EMPTY_ANSWER = { issues: [] };
class SonarTsServer {
    constructor() {
        this.fileCache = new languageService_1.FileCache();
        // key is ts file path, value is corresponding tsconfig path
        this.tsConfigCache = new Map();
        this.servicesPerTsconfig = new Map();
    }
    start(port) {
        logTypeScriptMetaInfo();
        const client = net.createConnection(port, "localhost", () => {
            console.log("SonarTS Server connected to " + port);
            let accumulatedData = "";
            client.on("data", data => {
                accumulatedData += data;
                try {
                    const request = JSON.parse(accumulatedData.toString());
                    accumulatedData = "";
                    const issues = this.handleAnalysisRequest(request);
                    client.write(JSON.stringify(issues));
                }
                catch (e) {
                    // ignore
                }
            });
        });
    }
    handleAnalysisRequest(request) {
        const { file, rules, content } = request;
        this.fileCache.newContent({ file, content });
        let tsConfig;
        if (this.tsConfigCache.has(file)) {
            tsConfig = this.tsConfigCache.get(file);
        }
        else {
            tsConfig = this.getTsConfig(file);
            if (tsConfig) {
                this.tsConfigCache.set(file, tsConfig);
            }
            else {
                console.error("No tsconfig.json file found for " + file);
                return EMPTY_ANSWER;
            }
        }
        let service = this.servicesPerTsconfig.get(tsConfig);
        if (!service) {
            const { files, options } = parser_1.parseTsConfig(tsConfig);
            service = languageService_1.createService(files, options, this.fileCache);
            this.servicesPerTsconfig.set(tsConfig, service);
        }
        const program = service.getProgram();
        const sourceFile = program.getSourceFile(file);
        if (!sourceFile) {
            console.error(`No SourceFile found for file ${file} with configuration ${tsConfig}`);
            return EMPTY_ANSWER;
        }
        const diagnostics = diagnostics_1.default(sourceFile, program);
        if (diagnostics.length > 0) {
            return { diagnostics };
        }
        return rules_1.getIssues(rules, program, sourceFile);
    }
    getTsConfig(filePath) {
        let currentDirectory = filePath;
        do {
            currentDirectory = path.dirname(currentDirectory);
            const possibleTsConfig = path.join(currentDirectory, "tsconfig.json");
            if (fs.existsSync(possibleTsConfig)) {
                return possibleTsConfig;
            }
        } while (currentDirectory !== path.dirname(currentDirectory));
        return undefined;
    }
}
exports.SonarTsServer = SonarTsServer;
function logTypeScriptMetaInfo() {
    const version = require("typescript/package.json").version;
    const location = path.dirname(path.dirname(require.resolve("typescript")));
    console.log(`Using typescript at [${location}], version ${version}`);
}
