// tslint-sonarts/lib/se/constraints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
var ConstraintKind;
(function (ConstraintKind) {
    ConstraintKind[ConstraintKind["Falsy"] = 0] = "Falsy";
    ConstraintKind[ConstraintKind["Truthy"] = 1] = "Truthy";
})(ConstraintKind = exports.ConstraintKind || (exports.ConstraintKind = {}));
function isTruthyConstraint(constraint) {
    return constraint.kind === ConstraintKind.Truthy;
}
exports.isTruthyConstraint = isTruthyConstraint;
function isFalsyConstraint(constraint) {
    return constraint.kind === ConstraintKind.Falsy;
}
exports.isFalsyConstraint = isFalsyConstraint;
function truthyConstraint() {
    return { kind: ConstraintKind.Truthy };
}
exports.truthyConstraint = truthyConstraint;
function falsyConstraint() {
    return { kind: ConstraintKind.Falsy };
}
exports.falsyConstraint = falsyConstraint;
function isEqualConstraints(a, b) {
    return a.kind === b.kind;
}
exports.isEqualConstraints = isEqualConstraints;
function constrain(list, candidate) {
    const newList = [];
    let existingKind = false;
    for (const item of list) {
        const result = constrainWith(item, candidate);
        if (!result) {
            return undefined;
        }
        newList.push(...result);
        if (item.kind === candidate.kind) {
            existingKind = true;
        }
    }
    if (!existingKind) {
        newList.push(candidate);
    }
    return newList;
}
exports.constrain = constrain;
function constrainWith(what, withWhat) {
    return what.kind === withWhat.kind ? [what] : undefined;
}
function isTruthy(constraints) {
    return constraints.length === 1 && isTruthyConstraint(constraints[0]);
}
exports.isTruthy = isTruthy;
function isFalsy(constraints) {
    return constraints.length === 1 && isFalsyConstraint(constraints[0]);
}
exports.isFalsy = isFalsy;
