// tslint-sonarts/lib/se/symbolicValues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
var SymbolicValueType;
(function (SymbolicValueType) {
    SymbolicValueType[SymbolicValueType["Unknown"] = 0] = "Unknown";
    SymbolicValueType[SymbolicValueType["NumericLiteral"] = 1] = "NumericLiteral";
    SymbolicValueType[SymbolicValueType["Undefined"] = 2] = "Undefined";
    SymbolicValueType[SymbolicValueType["ObjectLiteral"] = 3] = "ObjectLiteral";
})(SymbolicValueType = exports.SymbolicValueType || (exports.SymbolicValueType = {}));
function numericLiteralSymbolicValue(value) {
    return { type: SymbolicValueType.NumericLiteral, value };
}
exports.numericLiteralSymbolicValue = numericLiteralSymbolicValue;
function simpleSymbolicValue() {
    return { type: SymbolicValueType.Unknown };
}
exports.simpleSymbolicValue = simpleSymbolicValue;
function undefinedSymbolicValue() {
    return { type: SymbolicValueType.Undefined };
}
exports.undefinedSymbolicValue = undefinedSymbolicValue;
function objectLiteralSymbolicValue() {
    return { type: SymbolicValueType.ObjectLiteral };
}
exports.objectLiteralSymbolicValue = objectLiteralSymbolicValue;
function isEqualSymbolicValues(a, b) {
    return isNumericLiteralSymbolicValue(a) && isNumericLiteralSymbolicValue(b) ? a.value === b.value : a.type === b.type;
}
exports.isEqualSymbolicValues = isEqualSymbolicValues;
function isNumericLiteralSymbolicValue(value) {
    return value.type === SymbolicValueType.NumericLiteral;
}
exports.isNumericLiteralSymbolicValue = isNumericLiteralSymbolicValue;
function isUndefinedSymbolcValue(value) {
    return value.type === SymbolicValueType.Undefined;
}
exports.isUndefinedSymbolcValue = isUndefinedSymbolcValue;
