// tslint-sonarts/lib/se/SymbolicExecution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const cfg_1 = require("../cfg/cfg");
const stateTransitions_1 = require("./stateTransitions");
const navigation_1 = require("../utils/navigation");
const nodes_1 = require("../utils/nodes");
const BLOCK_VISITS_LIMIT = 1000;
function execute(cfg, symbols, initialState, shouldTrackSymbol = () => true) {
    const programNodes = new Map();
    const branchingProgramNodes = new Map();
    let visits = 0;
    visitBlock(cfg.start, initialState);
    if (visitsLimitBreached()) {
        // Analysis incomplete, it's safer to not report partial results
        return undefined;
    }
    else {
        return { programNodes, branchingProgramNodes, visits };
    }
    function visitBlock(block, programState) {
        if (programBlockAlreadyVisited(block, programState)) {
            return;
        }
        if (visitsLimitBreached()) {
            return;
        }
        visits++;
        for (const programPoint of block.getElements()) {
            programState = visitProgramPoint(programPoint, programState);
        }
        visitSuccessors(block, programState);
    }
    function programBlockAlreadyVisited(block, programState) {
        if (block instanceof cfg_1.CfgBranchingBlock && block.getElements().length > 0) {
            const existingStates = programNodes.get(block.getElements()[0]) || [];
            if (existingStates.find(existingState => programState.isEqualTo(existingState))) {
                return true;
            }
        }
        return false;
    }
    function visitSuccessors(block, programState) {
        // ignore for-of, for-in and switch, because we can't constrain right element
        if (block instanceof cfg_1.CfgBranchingBlock && !isForInOfLoop(block) && !isSwitch(block)) {
            addToBranchingNodes(block, programState);
            visitBranchingBlock(block, programState);
        }
        else {
            for (const successor of block.getSuccessors()) {
                visitBlock(successor, programState);
            }
        }
    }
    function visitBranchingBlock(block, programState) {
        let truthyState = programState.constrainToTruthy();
        let falsyState = programState.constrainToFalsy();
        // if we are inside a branching statement, remove its expression from the stack
        if (isStatement(block)) {
            truthyState = truthyState && truthyState.popSV()[1];
            falsyState = falsyState && falsyState.popSV()[1];
        }
        if (truthyState) {
            visitBlock(block.getTrueSuccessor(), truthyState);
        }
        if (falsyState) {
            visitBlock(block.getFalseSuccessor(), falsyState);
        }
    }
    function visitsLimitBreached() {
        return visits >= BLOCK_VISITS_LIMIT;
    }
    function visitProgramPoint(programPoint, programState) {
        addToProgramNodes(programPoint, programState);
        return stateTransitions_1.applyExecutors(programPoint, programState, symbols, shouldTrackSymbol);
    }
    function addToProgramNodes(programPoint, programState) {
        const visitedStates = programNodes.get(programPoint) || [];
        const nextVisitedStates = [...visitedStates, programState];
        programNodes.set(programPoint, nextVisitedStates);
    }
    function addToBranchingNodes(block, programState) {
        const lastElement = block.getElements()[block.getElements().length - 1];
        const existingStates = branchingProgramNodes.get(lastElement) || [];
        branchingProgramNodes.set(lastElement, [...existingStates, programState]);
    }
    function isForInOfLoop(block) {
        return nodes_1.is(block.loopingStatement, ts.SyntaxKind.ForInStatement, ts.SyntaxKind.ForOfStatement);
    }
    function isStatement(block) {
        return nodes_1.is(block.branchingElement, ...navigation_1.CONDITIONAL_STATEMENTS, ...navigation_1.LOOP_STATEMENTS);
    }
    function isSwitch(block) {
        return nodes_1.is(block.branchingElement, ts.SyntaxKind.SwitchStatement);
    }
}
exports.execute = execute;
