// tslint-sonarts/lib/utils/cognitiveComplexity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const visitor_1 = require("../utils/visitor");
const navigation_1 = require("../utils/navigation");
const nodes_1 = require("../utils/nodes");
function getOverallCognitiveComplexity(node) {
    let complexityWalker = new ComplexityWalker();
    complexityWalker.visit(node);
    let fileComplexity = complexityWalker.getComplexity().complexity;
    const functionCollector = new FunctionCollector();
    functionCollector.visit(node);
    functionCollector.functionComplexities.forEach(functionComplexity => (fileComplexity += functionComplexity.complexity));
    return fileComplexity;
}
exports.getOverallCognitiveComplexity = getOverallCognitiveComplexity;
function getFunctionCognitiveComplexity(node) {
    let complexityWalker = new ComplexityWalker();
    complexityWalker.visit(node.body);
    return complexityWalker.getComplexity();
}
const TARGETED_KINDS = [
    // ArrowFunctions and FunctionExpressions are included in computation
    ts.SyntaxKind.FunctionDeclaration,
    ts.SyntaxKind.MethodDeclaration,
    ts.SyntaxKind.Constructor,
    ts.SyntaxKind.GetAccessor,
    ts.SyntaxKind.SetAccessor,
];
class FunctionCollector extends visitor_1.TreeVisitor {
    constructor() {
        super(...arguments);
        this.functionComplexities = [];
    }
    visitNode(node) {
        if (nodes_1.is(node, ...TARGETED_KINDS)) {
            const functionNode = node;
            const body = functionNode.body;
            if (body) {
                const complexity = getFunctionCognitiveComplexity(node);
                this.functionComplexities.push(Object.assign({ functionNode }, complexity));
            }
        }
        super.visitNode(node);
    }
}
exports.FunctionCollector = FunctionCollector;
class ComplexityWalker extends visitor_1.TreeVisitor {
    constructor() {
        super(...arguments);
        this.complexity = 0;
        this.complexityNodes = [];
        this.nesting = 0;
        this.logicalOperationsToIgnore = [];
    }
    getComplexity() {
        return { complexity: this.complexity, nodes: this.complexityNodes };
    }
    visitNode(node) {
        if (!nodes_1.is(node, ...TARGETED_KINDS)) {
            super.visitNode(node);
        }
    }
    visitIfStatement(node) {
        const ifToken = navigation_1.findChild(node, ts.SyntaxKind.IfKeyword);
        if (this.isElseIf(node)) {
            this.addComplexityWithoutNesting(ifToken);
        }
        else {
            this.addComplexityWithNesting(ifToken);
        }
        this.visit(node.expression);
        this.visitWithNesting(node.thenStatement);
        if (node.elseStatement) {
            if (nodes_1.is(node.elseStatement, ts.SyntaxKind.IfStatement)) {
                this.visit(node.elseStatement);
            }
            else {
                this.addComplexityWithoutNesting(navigation_1.findChild(node, ts.SyntaxKind.ElseKeyword));
                this.visitWithNesting(node.elseStatement);
            }
        }
    }
    isElseIf(node) {
        if (node.parent) {
            return nodes_1.is(node.parent, ts.SyntaxKind.IfStatement) && node === node.parent.elseStatement;
        }
        return false;
    }
    visitSwitchStatement(node) {
        this.addComplexityWithNesting(navigation_1.findChild(node, ts.SyntaxKind.SwitchKeyword));
        this.visit(node.expression);
        this.visitWithNesting(node.caseBlock);
    }
    visitFunctionExpression(node) {
        this.visitWithNesting(node.body);
    }
    visitArrowFunction(node) {
        this.visitWithNesting(node.body);
    }
    visitContinueStatement(node) {
        if (node.label) {
            this.addComplexityWithoutNesting(navigation_1.findChild(node, ts.SyntaxKind.ContinueKeyword));
        }
        super.visitContinueStatement(node);
    }
    visitBreakStatement(node) {
        if (node.label) {
            this.addComplexityWithoutNesting(navigation_1.findChild(node, ts.SyntaxKind.BreakKeyword));
        }
        super.visitBreakStatement(node);
    }
    visitCatchClause(node) {
        this.addComplexityWithNesting(navigation_1.findChild(node, ts.SyntaxKind.CatchKeyword));
        this.visitWithNesting(node.block);
    }
    visitForStatement(node) {
        this.visitLoop(node.statement, ts.SyntaxKind.ForKeyword, node.initializer, node.condition, node.incrementor);
    }
    visitForInStatement(node) {
        this.visitLoop(node.statement, ts.SyntaxKind.ForKeyword, node.initializer, node.expression);
    }
    visitForOfStatement(node) {
        this.visitLoop(node.statement, ts.SyntaxKind.ForKeyword, node.initializer, node.expression);
    }
    visitDoStatement(node) {
        this.visitLoop(node.statement, ts.SyntaxKind.DoKeyword, node.expression);
    }
    visitWhileStatement(node) {
        this.visitLoop(node.statement, ts.SyntaxKind.WhileKeyword, node.expression);
    }
    visitLoop(nestedNode, complexityKeyword, ...notNestedExpressions) {
        this.addComplexityWithNesting(navigation_1.findChild(nestedNode.parent, complexityKeyword));
        this.walkNodes(notNestedExpressions);
        this.visitWithNesting(nestedNode);
    }
    visitConditionalExpression(node) {
        this.addComplexityWithNesting(navigation_1.findChild(node, ts.SyntaxKind.QuestionToken));
        this.visit(node.condition);
        this.visitWithNesting(node.whenTrue);
        this.visitWithNesting(node.whenFalse);
    }
    visitBinaryExpression(node) {
        if (ComplexityWalker.isAndOrOr(node) && !this.isIgnoredOperation(node)) {
            const flattenedLogicalExpressions = this.flattenLogicalExpression(node);
            let previous;
            for (const current of flattenedLogicalExpressions) {
                if (!previous || !nodes_1.is(previous.operatorToken, current.operatorToken.kind)) {
                    this.addComplexityWithoutNesting(current.operatorToken);
                }
                previous = current;
            }
        }
        super.visitBinaryExpression(node);
    }
    static isAndOrOr(node) {
        if (nodes_1.is(node, ts.SyntaxKind.BinaryExpression)) {
            return nodes_1.is(node.operatorToken, ts.SyntaxKind.AmpersandAmpersandToken, ts.SyntaxKind.BarBarToken);
        }
        return false;
    }
    flattenLogicalExpression(node) {
        if (ComplexityWalker.isAndOrOr(node)) {
            const binaryExpression = node;
            this.logicalOperationsToIgnore.push(binaryExpression);
            const leftChild = navigation_1.drillDownThroughParenthesis(binaryExpression.left);
            const rightChild = navigation_1.drillDownThroughParenthesis(binaryExpression.right);
            return this.flattenLogicalExpression(leftChild)
                .concat([binaryExpression])
                .concat(this.flattenLogicalExpression(rightChild));
        }
        return [];
    }
    isIgnoredOperation(node) {
        return this.logicalOperationsToIgnore.indexOf(node) > -1;
    }
    walkNodes(nodes) {
        for (const node of nodes) {
            if (node) {
                this.visit(node);
            }
        }
    }
    visitWithNesting(node) {
        this.nesting++;
        this.visit(node);
        this.nesting--;
    }
    addComplexityWithoutNesting(node) {
        this.complexity += 1;
        this.complexityNodes.push({ node, complexity: 1 });
    }
    addComplexityWithNesting(node, value) {
        const addition = value ? value : 1;
        const nodeComplexity = this.nesting + addition;
        this.complexity += nodeComplexity;
        this.complexityNodes.push({ node, complexity: nodeComplexity });
    }
}
