// tslint-sonarts/lib/utils/navigation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const nodes = require("./nodes");
const Kind = ts.SyntaxKind;
function keyword(node) {
    return node.getFirstToken();
}
exports.keyword = keyword;
function collectLeftHandIdentifiers(node) {
    const identifiers = [];
    const nonIdentifiers = [];
    collectFromExpression(node);
    identifiers.reverse();
    nonIdentifiers.reverse();
    return { identifiers, nonIdentifiers };
    function collectFromExpression(node) {
        node = drillDownThroughParenthesis(node);
        if (nodes.isIdentifier(node)) {
            identifiers.push(node);
        }
        else if (nodes.isObjectLiteralExpression(node)) {
            collectFromObjectLiteralExpression(node);
        }
        else if (nodes.isArrayLiteralExpression(node)) {
            node.elements.forEach(element => collectFromExpression(element));
        }
        else if (nodes.isSpreadElement(node)) {
            collectFromExpression(node.expression);
        }
        else if (nodes.isBinaryExpression(node)) {
            collectFromExpression(node.left);
            nonIdentifiers.push(node.right);
        }
        else {
            nonIdentifiers.push(node);
        }
    }
    function collectFromObjectLiteralExpression(node) {
        node.properties.forEach(property => {
            if (nodes.isPropertyAssignment(property)) {
                collectFromExpression(property.initializer);
            }
            else if (nodes.isShorthandPropertyAssignment(property)) {
                collectFromExpression(property.name);
                if (property.objectAssignmentInitializer) {
                    nonIdentifiers.push(property.objectAssignmentInitializer);
                }
            }
            else if (nodes.isSpreadAssignment(property)) {
                collectFromExpression(property.expression);
            }
        });
    }
}
exports.collectLeftHandIdentifiers = collectLeftHandIdentifiers;
function getComments(node) {
    return [...getCommentsBefore(node), ...getCommentsAfter(node)];
}
exports.getComments = getComments;
function getCommentsBefore(node) {
    return ts.getLeadingCommentRanges(node.getSourceFile().text, node.getFullStart()) || [];
}
exports.getCommentsBefore = getCommentsBefore;
function getCommentsAfter(node) {
    return ts.getTrailingCommentRanges(node.getSourceFile().text, node.getEnd()) || [];
}
exports.getCommentsAfter = getCommentsAfter;
function getText(textRange, file) {
    return file.getFullText().substr(textRange.pos, textRange.end - textRange.pos);
}
exports.getText = getText;
function toTokens(node) {
    const result = [];
    const stack = [node];
    while (stack.length) {
        const currentNode = stack.pop();
        if (nodes.isToken(currentNode)) {
            result.push(currentNode);
            continue;
        }
        // skip jsDoc
        if (currentNode.kind >= ts.SyntaxKind.FirstJSDocNode && currentNode.kind <= ts.SyntaxKind.LastJSDocNode) {
            continue;
        }
        stack.push(...currentNode.getChildren());
    }
    return result.reverse();
}
exports.toTokens = toTokens;
function lineAndCharacterByPos(pos, file) {
    return file.getLineAndCharacterOfPosition(pos);
}
exports.lineAndCharacterByPos = lineAndCharacterByPos;
function startLineAndCharacter(node) {
    return node.getSourceFile().getLineAndCharacterOfPosition(node.getStart());
}
exports.startLineAndCharacter = startLineAndCharacter;
function endLineAndCharacter(node) {
    return node.getSourceFile().getLineAndCharacterOfPosition(node.getEnd());
}
exports.endLineAndCharacter = endLineAndCharacter;
function localAncestorsChain(node) {
    return ancestorsChain(node, ...exports.FUNCTION_LIKE);
}
exports.localAncestorsChain = localAncestorsChain;
function ancestorsChain(node, ...boundary) {
    const chain = [];
    for (let parent = node.parent; !!parent; parent = parent.parent) {
        chain.push(parent);
        if (boundary.includes(parent.kind))
            break;
    }
    return chain;
}
exports.ancestorsChain = ancestorsChain;
function firstLocalAncestor(node, ...targetAncestor) {
    return firstAncestor(node, targetAncestor, ...exports.FUNCTION_LIKE);
}
exports.firstLocalAncestor = firstLocalAncestor;
function firstAncestor(node, targetAncestor, ...boundary) {
    return ancestorsChain(node, ...boundary).find(ancestor => targetAncestor.includes(ancestor.kind));
}
exports.firstAncestor = firstAncestor;
function floatToTopParenthesis(node) {
    return nodes.isParenthesizedExpression(node) && node.parent ? floatToTopParenthesis(node.parent) : node;
}
exports.floatToTopParenthesis = floatToTopParenthesis;
function drillDownThroughParenthesis(node) {
    return nodes.isParenthesizedExpression(node) ? drillDownThroughParenthesis(node.expression) : node;
}
exports.drillDownThroughParenthesis = drillDownThroughParenthesis;
/** Returns all descendants of the `node`, including tokens */
function descendants(node) {
    const children = node.getChildren();
    let collectedDescendants = children;
    children.forEach(child => (collectedDescendants = collectedDescendants.concat(descendants(child))));
    return collectedDescendants;
}
exports.descendants = descendants;
function findChild(node, kind) {
    const child = node.getChildren().find(child => nodes.is(child, kind));
    if (child) {
        return child;
    }
    else {
        throw new Error("Not found child with kind " + ts.SyntaxKind[kind]);
    }
}
exports.findChild = findChild;
function accessModifier(declaration) {
    if (declaration.modifiers) {
        return declaration.modifiers.find(modifier => nodes.is(modifier, ...exports.ACCESS_MODIFIERS));
    }
    else {
        return;
    }
}
exports.accessModifier = accessModifier;
function isReadonly(declaration) {
    if (declaration.modifiers) {
        return declaration.modifiers.find(modifier => nodes.is(modifier, ts.SyntaxKind.ReadonlyKeyword));
    }
    else {
        return;
    }
}
exports.isReadonly = isReadonly;
function constructorOf(clazz) {
    return clazz.members.find(member => member.kind === ts.SyntaxKind.Constructor);
}
exports.constructorOf = constructorOf;
/**
 * Returns
 * - function name token for methods and accessors
 * - "function" keyword for function declarations and expressions
 * - "=>" for arrow function
 */
function functionLikeMainToken(functionNode) {
    switch (functionNode.kind) {
        case Kind.FunctionDeclaration:
        case Kind.FunctionExpression:
            return getFirstChild(functionNode, ts.SyntaxKind.FunctionKeyword);
        case Kind.ArrowFunction:
            return functionNode.equalsGreaterThanToken;
        case Kind.Constructor:
            return getFirstChild(functionNode, ts.SyntaxKind.ConstructorKeyword);
        case Kind.MethodDeclaration:
        case Kind.GetAccessor:
        case Kind.SetAccessor:
            return functionNode.name;
    }
}
exports.functionLikeMainToken = functionLikeMainToken;
function getFirstChild(node, childKind) {
    return node.getChildren().find(child => child.kind === childKind);
}
exports.getFirstChild = getFirstChild;
exports.FUNCTION_LIKE = [
    Kind.FunctionDeclaration,
    Kind.FunctionExpression,
    Kind.ArrowFunction,
    Kind.MethodDeclaration,
    Kind.Constructor,
    Kind.GetAccessor,
    Kind.SetAccessor,
];
exports.CONDITIONAL_STATEMENTS = [Kind.IfStatement, Kind.SwitchStatement];
exports.LOOP_STATEMENTS = [
    Kind.ForStatement,
    Kind.ForInStatement,
    Kind.ForOfStatement,
    Kind.WhileStatement,
    Kind.DoStatement,
];
exports.COMPOUND_ASSIGNMENTS = [
    Kind.PlusEqualsToken,
    Kind.MinusEqualsToken,
    Kind.AsteriskAsteriskEqualsToken,
    Kind.AsteriskEqualsToken,
    Kind.SlashEqualsToken,
    Kind.PercentEqualsToken,
    Kind.AmpersandEqualsToken,
    Kind.BarEqualsToken,
    Kind.CaretEqualsToken,
    Kind.LessThanLessThanEqualsToken,
    Kind.GreaterThanGreaterThanGreaterThanEqualsToken,
    Kind.GreaterThanGreaterThanEqualsToken,
];
exports.ACCESS_MODIFIERS = [Kind.PublicKeyword, Kind.PrivateKeyword, Kind.ProtectedKeyword];
