// tslint-sonarts/lib/utils/nodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const navigation_1 = require("./navigation");
// Note: keep all type guards ordered aphabetically
function is(node, ...kinds) {
    return node !== undefined && kinds.includes(node.kind);
}
exports.is = is;
function isArrayBindingPattern(node) {
    return is(node, ts.SyntaxKind.ArrayBindingPattern);
}
exports.isArrayBindingPattern = isArrayBindingPattern;
function isArrayLiteralExpression(node) {
    return is(node, ts.SyntaxKind.ArrayLiteralExpression);
}
exports.isArrayLiteralExpression = isArrayLiteralExpression;
function isArrowFunction(node) {
    return is(node, ts.SyntaxKind.ArrowFunction);
}
exports.isArrowFunction = isArrowFunction;
function isAssignment(node) {
    return isBinaryExpression(node) && node.operatorToken.kind === ts.SyntaxKind.EqualsToken;
}
exports.isAssignment = isAssignment;
function isAssignmentKind(kind) {
    return kind >= ts.SyntaxKind.FirstAssignment && kind <= ts.SyntaxKind.LastAssignment;
}
exports.isAssignmentKind = isAssignmentKind;
function isBinaryExpression(node) {
    return is(node, ts.SyntaxKind.BinaryExpression);
}
exports.isBinaryExpression = isBinaryExpression;
function isBindingElement(node) {
    return is(node, ts.SyntaxKind.BindingElement);
}
exports.isBindingElement = isBindingElement;
function isBlock(node) {
    return is(node, ts.SyntaxKind.Block);
}
exports.isBlock = isBlock;
function isBreakStatement(node) {
    return is(node, ts.SyntaxKind.BreakStatement);
}
exports.isBreakStatement = isBreakStatement;
function isCallExpression(node) {
    return is(node, ts.SyntaxKind.CallExpression);
}
exports.isCallExpression = isCallExpression;
function isCaseClause(node) {
    return is(node, ts.SyntaxKind.CaseClause);
}
exports.isCaseClause = isCaseClause;
function isCatchClause(node) {
    return is(node, ts.SyntaxKind.CatchClause);
}
exports.isCatchClause = isCatchClause;
function isContinueStatement(node) {
    return is(node, ts.SyntaxKind.ContinueStatement);
}
exports.isContinueStatement = isContinueStatement;
function isElementAccessExpression(node) {
    return is(node, ts.SyntaxKind.ElementAccessExpression);
}
exports.isElementAccessExpression = isElementAccessExpression;
function isExpressionStatement(node) {
    return is(node, ts.SyntaxKind.ExpressionStatement);
}
exports.isExpressionStatement = isExpressionStatement;
function isFunctionDeclaration(node) {
    return is(node, ts.SyntaxKind.FunctionDeclaration);
}
exports.isFunctionDeclaration = isFunctionDeclaration;
function isFunctionExpression(node) {
    return is(node, ts.SyntaxKind.FunctionExpression);
}
exports.isFunctionExpression = isFunctionExpression;
function isFunctionLikeDeclaration(node) {
    return is(node, ...navigation_1.FUNCTION_LIKE);
}
exports.isFunctionLikeDeclaration = isFunctionLikeDeclaration;
function isIdentifier(node) {
    return is(node, ts.SyntaxKind.Identifier);
}
exports.isIdentifier = isIdentifier;
function isInterfaceDeclaration(node) {
    return is(node, ts.SyntaxKind.InterfaceDeclaration);
}
exports.isInterfaceDeclaration = isInterfaceDeclaration;
function isIfStatement(node) {
    return is(node, ts.SyntaxKind.IfStatement);
}
exports.isIfStatement = isIfStatement;
function isIterationStatement(node) {
    return is(node, ts.SyntaxKind.WhileStatement, ts.SyntaxKind.DoStatement, ts.SyntaxKind.ForStatement, ts.SyntaxKind.ForInStatement, ts.SyntaxKind.ForOfStatement);
}
exports.isIterationStatement = isIterationStatement;
function isLabeledStatement(node) {
    return is(node, ts.SyntaxKind.LabeledStatement);
}
exports.isLabeledStatement = isLabeledStatement;
function isLiteralExpression(node) {
    return ts.SyntaxKind.FirstLiteralToken <= node.kind && node.kind <= ts.SyntaxKind.LastLiteralToken;
}
exports.isLiteralExpression = isLiteralExpression;
function isNewExpression(node) {
    return is(node, ts.SyntaxKind.NewExpression);
}
exports.isNewExpression = isNewExpression;
function isNumericLiteral(node) {
    return is(node, ts.SyntaxKind.NumericLiteral);
}
exports.isNumericLiteral = isNumericLiteral;
function isObjectBindingPattern(node) {
    return is(node, ts.SyntaxKind.ObjectBindingPattern);
}
exports.isObjectBindingPattern = isObjectBindingPattern;
function isObjectLiteralExpression(node) {
    return is(node, ts.SyntaxKind.ObjectLiteralExpression);
}
exports.isObjectLiteralExpression = isObjectLiteralExpression;
function isParenthesizedExpression(node) {
    return is(node, ts.SyntaxKind.ParenthesizedExpression);
}
exports.isParenthesizedExpression = isParenthesizedExpression;
function isPostfixUnaryExpression(node) {
    return is(node, ts.SyntaxKind.PostfixUnaryExpression);
}
exports.isPostfixUnaryExpression = isPostfixUnaryExpression;
function isPrefixUnaryExpression(node) {
    return is(node, ts.SyntaxKind.PrefixUnaryExpression);
}
exports.isPrefixUnaryExpression = isPrefixUnaryExpression;
function isPropertyAccessExpression(node) {
    return is(node, ts.SyntaxKind.PropertyAccessExpression);
}
exports.isPropertyAccessExpression = isPropertyAccessExpression;
function isPropertyAssignment(node) {
    return is(node, ts.SyntaxKind.PropertyAssignment);
}
exports.isPropertyAssignment = isPropertyAssignment;
function isReturnStatement(node) {
    return is(node, ts.SyntaxKind.ReturnStatement);
}
exports.isReturnStatement = isReturnStatement;
function isShorthandPropertyAssignment(node) {
    return is(node, ts.SyntaxKind.ShorthandPropertyAssignment);
}
exports.isShorthandPropertyAssignment = isShorthandPropertyAssignment;
function isSpreadAssignment(node) {
    return is(node, ts.SyntaxKind.SpreadAssignment);
}
exports.isSpreadAssignment = isSpreadAssignment;
function isSpreadElement(node) {
    return is(node, ts.SyntaxKind.SpreadElement);
}
exports.isSpreadElement = isSpreadElement;
function isStringLiteral(node) {
    return is(node, ts.SyntaxKind.StringLiteral);
}
exports.isStringLiteral = isStringLiteral;
function isToken(node) {
    return node.kind <= ts.SyntaxKind.OfKeyword;
}
exports.isToken = isToken;
function isThrowStatement(node) {
    return is(node, ts.SyntaxKind.ThrowStatement);
}
exports.isThrowStatement = isThrowStatement;
function isTypeNode(node) {
    return ((node.kind >= ts.SyntaxKind.FirstTypeNode && node.kind <= ts.SyntaxKind.LastTypeNode) ||
        node.kind === ts.SyntaxKind.AnyKeyword ||
        node.kind === ts.SyntaxKind.NumberKeyword ||
        node.kind === ts.SyntaxKind.ObjectKeyword ||
        node.kind === ts.SyntaxKind.BooleanKeyword ||
        node.kind === ts.SyntaxKind.StringKeyword ||
        node.kind === ts.SyntaxKind.SymbolKeyword ||
        node.kind === ts.SyntaxKind.ThisKeyword ||
        node.kind === ts.SyntaxKind.VoidKeyword ||
        node.kind === ts.SyntaxKind.UndefinedKeyword ||
        node.kind === ts.SyntaxKind.NullKeyword ||
        node.kind === ts.SyntaxKind.NeverKeyword ||
        node.kind === ts.SyntaxKind.ExpressionWithTypeArguments ||
        node.kind === ts.SyntaxKind.JSDocAllType ||
        node.kind === ts.SyntaxKind.JSDocUnknownType ||
        node.kind === ts.SyntaxKind.JSDocNullableType ||
        node.kind === ts.SyntaxKind.JSDocNonNullableType ||
        node.kind === ts.SyntaxKind.JSDocOptionalType ||
        node.kind === ts.SyntaxKind.JSDocFunctionType ||
        node.kind === ts.SyntaxKind.JSDocVariadicType);
}
exports.isTypeNode = isTypeNode;
function isVariableDeclaration(node) {
    return is(node, ts.SyntaxKind.VariableDeclaration);
}
exports.isVariableDeclaration = isVariableDeclaration;
function isVariableDeclarationList(node) {
    return is(node, ts.SyntaxKind.VariableDeclarationList);
}
exports.isVariableDeclarationList = isVariableDeclarationList;
function isVariableStatement(node) {
    return is(node, ts.SyntaxKind.VariableStatement);
}
exports.isVariableStatement = isVariableStatement;
function isUnionOrIntersectionTypeNode(node) {
    return is(node, ts.SyntaxKind.UnionType, ts.SyntaxKind.IntersectionType);
}
exports.isUnionOrIntersectionTypeNode = isUnionOrIntersectionTypeNode;
function isSignatureDeclaration(node) {
    return "parameters" in node;
}
exports.isSignatureDeclaration = isSignatureDeclaration;
