// tslint-sonarts/lib/utils/sonarAnalysis.js
"use strict";
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const navigation_1 = require("./navigation");
const visitor_1 = require("./visitor");
const tslint = require("tslint");
class SonarRuleVisitor extends visitor_1.TreeVisitor {
    constructor(ruleName) {
        super();
        this.ruleName = ruleName;
        this.issues = [];
    }
    getIssues() {
        return this.issues;
    }
    addIssue(node, message) {
        const issue = new SonarIssue(getIssueLocationAtNode(node, message), this.ruleName);
        this.issues.push(issue);
        return issue;
    }
    addIssueAtLocation(primaryLocation) {
        const issue = new SonarIssue(primaryLocation, this.ruleName);
        this.issues.push(issue);
        return issue;
    }
}
exports.SonarRuleVisitor = SonarRuleVisitor;
class TypedSonarRuleVisitor extends SonarRuleVisitor {
    constructor(ruleName, program) {
        super(ruleName);
        this.program = program;
    }
}
exports.TypedSonarRuleVisitor = TypedSonarRuleVisitor;
class IssueLocation {
    constructor(start, end, sourceFile, message) {
        this.start = start;
        this.end = end;
        this.sourceFile = sourceFile;
        this.message = message;
        this.message = message;
        const startPosition = navigation_1.lineAndCharacterByPos(start, sourceFile);
        const endPosition = navigation_1.lineAndCharacterByPos(end, sourceFile);
        this.startLine = startPosition.line;
        this.startColumn = startPosition.character;
        this.endLine = endPosition.line;
        this.endColumn = endPosition.character;
    }
    toJson() {
        return {
            startLine: this.startLine,
            startCol: this.startColumn,
            endLine: this.endLine,
            endCol: this.endColumn,
            message: this.message,
        };
    }
}
exports.IssueLocation = IssueLocation;
function getIssueLocationAtNode(node, message) {
    return new IssueLocation(node.getStart(), node.getEnd(), node.getSourceFile(), message);
}
exports.getIssueLocationAtNode = getIssueLocationAtNode;
class SonarIssue extends tslint.RuleFailure {
    constructor(primaryLocation, ruleName) {
        super(primaryLocation.sourceFile, primaryLocation.start, primaryLocation.end, primaryLocation.message || "", ruleName);
        this.secondaryLocations = [];
        this.primaryLocation = primaryLocation;
    }
    toJson() {
        return {
            failure: this.primaryLocation.message,
            startPosition: {
                line: this.primaryLocation.startLine,
                character: this.primaryLocation.startColumn,
                position: this.primaryLocation.start,
            },
            endPosition: {
                line: this.primaryLocation.endLine,
                character: this.primaryLocation.endColumn,
                position: this.primaryLocation.end,
            },
            name: this.primaryLocation.sourceFile.fileName,
            ruleName: this.getRuleName(),
            cost: this.cost,
            secondaryLocations: this.secondaryLocations.map(location => location.toJson()),
            ruleSeverity: this.getRuleSeverity(),
        };
    }
    setCost(cost) {
        this.cost = cost;
        return this;
    }
    addSecondaryLocation(nodeOrLocation, message) {
        if (nodeOrLocation instanceof IssueLocation) {
            this.secondaryLocations.push(nodeOrLocation);
        }
        else {
            const issueLocation = getIssueLocationAtNode(nodeOrLocation);
            issueLocation.message = message;
            this.addSecondaryLocation(issueLocation);
        }
        return this;
    }
    getSecondaryLocations() {
        return this.secondaryLocations;
    }
    getCost() {
        return this.cost;
    }
}
exports.SonarIssue = SonarIssue;
