// tslint-sonarts/lib/symbols/builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
const table_1 = require("./table");
const visitor_1 = require("../utils/visitor");
const nodes_1 = require("../utils/nodes");
class SymbolTableBuilder extends visitor_1.TreeVisitor {
    constructor(program) {
        super();
        this.program = program;
        this.table = new table_1.SymbolTable();
    }
    static build(sourceFile, program) {
        if (!sourceFile.symbolTable) {
            const builder = new SymbolTableBuilder(program);
            builder.visit(sourceFile);
            sourceFile.symbolTable = builder.table;
        }
        return sourceFile.symbolTable;
    }
    visitIdentifier(node) {
        this.registerUsageIfMissing(node, table_1.UsageFlag.READ);
    }
    visitBinaryExpression(node) {
        if (nodes_1.isAssignment(node)) {
            this.registerWriteUsageForAssignment(node.left);
        }
        super.visitBinaryExpression(node);
    }
    visitVariableDeclaration(node) {
        this.addVariable(node);
        super.visitVariableDeclaration(node);
    }
    visitPropertyDeclaration(node) {
        this.addVariable(node);
        super.visitPropertyDeclaration(node);
    }
    visitParameterDeclaration(node) {
        this.addVariable(node);
        super.visitParameterDeclaration(node);
    }
    visitFunctionDeclaration(node) {
        if (node.name)
            this.registerUsageIfMissing(node.name, table_1.UsageFlag.DECLARATION);
        super.visitFunctionDeclaration(node);
    }
    visitClassDeclaration(node) {
        if (node.name)
            this.registerUsageIfMissing(node.name, table_1.UsageFlag.DECLARATION);
        super.visitClassDeclaration(node);
    }
    visitEnumDeclaration(node) {
        if (node.name)
            this.registerUsageIfMissing(node.name, table_1.UsageFlag.DECLARATION);
        super.visitEnumDeclaration(node);
    }
    visitInterfaceDeclaration(node) {
        if (node.name)
            this.registerUsageIfMissing(node.name, table_1.UsageFlag.DECLARATION);
        super.visitInterfaceDeclaration(node);
    }
    visitNamedImports(node) {
        node.elements.forEach(importSpecifier => this.registerUsageIfMissing(importSpecifier.name, table_1.UsageFlag.DECLARATION));
        super.visitNamedImports(node);
    }
    visitNamespaceImport(node) {
        this.registerUsageIfMissing(node.name, table_1.UsageFlag.DECLARATION);
        super.visitNamespaceImport(node);
    }
    visitImportEqualsDeclaration(node) {
        this.registerUsageIfMissing(node.name, table_1.UsageFlag.DECLARATION);
        super.visitImportEqualsDeclaration(node);
    }
    visitExportAssignment(node) {
        // TODO This doesn't seem to intercept 'export let x = 42'
        super.visitExportAssignment(node);
    }
    visitPrefixUnaryExpression(node) {
        if (node.operator === ts.SyntaxKind.PlusPlusToken || node.operator === ts.SyntaxKind.MinusMinusToken) {
            this.registerUsageIfMissing(node.operand, table_1.UsageFlag.READ | table_1.UsageFlag.WRITE);
        }
        super.visitPrefixUnaryExpression(node);
    }
    visitPostfixUnaryExpression(node) {
        this.registerUsageIfMissing(node.operand, table_1.UsageFlag.READ | table_1.UsageFlag.WRITE);
    }
    visitModuleDeclaration(node) {
        this.registerUsageIfMissing(node.name, table_1.UsageFlag.DECLARATION);
        super.visitModuleDeclaration(node);
    }
    registerWriteUsageForAssignment(node) {
        navigation_1.collectLeftHandIdentifiers(node).identifiers.forEach(identifier => this.registerUsageIfMissing(identifier, table_1.UsageFlag.WRITE));
    }
    addVariable(node) {
        const declarationName = node.name;
        if (nodes_1.isIdentifier(declarationName)) {
            let usageFlags = table_1.UsageFlag.DECLARATION;
            if (node.initializer ||
                nodes_1.is(node, ts.SyntaxKind.Parameter) ||
                nodes_1.is(node.parent, ts.SyntaxKind.ObjectBindingPattern, ts.SyntaxKind.ArrayBindingPattern)) {
                usageFlags += table_1.UsageFlag.WRITE;
            }
            this.registerUsageIfMissing(declarationName, usageFlags);
        }
        else if (nodes_1.isArrayBindingPattern(declarationName) || nodes_1.isObjectBindingPattern(declarationName)) {
            Array.from(declarationName.elements).forEach(element => {
                if (nodes_1.isBindingElement(element)) {
                    this.addVariable(element);
                }
            });
        }
    }
    registerUsageIfMissing(node, flags) {
        if (nodes_1.isParenthesizedExpression(node)) {
            this.registerUsageIfMissing(node.expression, flags);
        }
        else {
            let symbol = this.program.getTypeChecker().getSymbolAtLocation(node);
            if (node.parent && node.parent.kind === ts.SyntaxKind.ShorthandPropertyAssignment) {
                symbol = this.program.getTypeChecker().getShorthandAssignmentValueSymbol(node.parent);
            }
            if (symbol)
                this.table.registerUsageIfMissing(symbol, node, flags);
        }
    }
}
exports.SymbolTableBuilder = SymbolTableBuilder;
