// tslint-sonarts/lib/symbols/lva.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const cfg_1 = require("../cfg/cfg");
const navigation_1 = require("../utils/navigation");
const table_1 = require("./table");
const nodes_1 = require("../utils/nodes");
class LiveVariableAnalyzer {
    constructor(symbols) {
        this.symbols = symbols;
        this.deadUsages = new Set();
        this.usedInNestedFunctionOrClass = new Map();
    }
    analyzeFunction(root) {
        if (root.body && nodes_1.isBlock(root.body)) {
            const cfg = cfg_1.ControlFlowGraph.fromStatements(Array.from(root.body.statements));
            return cfg && this.analyze(root, cfg);
        }
    }
    analyze(root, cfg) {
        this.root = (nodes_1.is(root, ...navigation_1.FUNCTION_LIKE)
            ? root
            : navigation_1.firstAncestor(root, LiveVariableAnalyzer.FUNCTION_OR_SOURCE_FILE));
        // symbols whose value will be read after entering a block (aka live symbols)
        this.blockAvailableReads = new Map();
        const blocks = cfg.getBlocks().concat(cfg.end);
        while (blocks.length > 0) {
            const block = blocks.pop();
            // live-in symbols from previous iteration of the algorithm for this block
            const oldBlockReads = this.blockAvailableReads.get(block);
            const newBlockReads = this.computeSymbolsWithAvailableReads(block);
            if (!this.same(newBlockReads, oldBlockReads)) {
                if (block instanceof cfg_1.CfgBlockWithPredecessors) {
                    blocks.unshift(...block.predecessors);
                }
            }
            this.blockAvailableReads.set(block, newBlockReads);
        }
        return { cfg, blockAvailableReads: this.blockAvailableReads, deadUsages: this.deadUsages };
    }
    computeSymbolsWithAvailableReads(block) {
        const availableReads = this.successorSymbolsWithAvailableReads(block);
        [...block.getElements()].reverse().forEach(element => {
            if (nodes_1.isAssignment(element)) {
                navigation_1.collectLeftHandIdentifiers(element.left).identifiers.forEach(identifier => {
                    this.trackUsage(this.symbols.getUsage(identifier), availableReads);
                });
            }
            else {
                this.trackUsage(this.symbols.getUsage(element), availableReads);
            }
        });
        return availableReads;
    }
    trackUsage(usage, availableReads) {
        if (usage && !this.isUsedInNestedFunctionOrClass(usage.symbol)) {
            if (usage.is(table_1.UsageFlag.WRITE)) {
                if (availableReads.has(usage.symbol)) {
                    this.deadUsages.delete(usage);
                    availableReads.delete(usage.symbol);
                }
                else {
                    this.deadUsages.add(usage);
                }
            }
            if (usage.is(table_1.UsageFlag.READ)) {
                availableReads.add(usage.symbol);
            }
        }
    }
    isUsedInNestedFunctionOrClass(symbol) {
        const cached = this.usedInNestedFunctionOrClass.get(symbol);
        if (cached !== undefined) {
            return cached;
        }
        const used = this.symbols
            .allUsages(symbol)
            .some(usage => navigation_1.firstLocalAncestor(usage.node, ...LiveVariableAnalyzer.FUNCTION_OR_SOURCE_FILE, ts.SyntaxKind.ClassDeclaration, ts.SyntaxKind.ClassExpression) !== this.root);
        this.usedInNestedFunctionOrClass.set(symbol, used);
        return used;
    }
    successorSymbolsWithAvailableReads(block) {
        const availableReads = new Set();
        block.getSuccessors().forEach(successor => {
            const availableReadsInSuccessor = this.blockAvailableReads.get(successor);
            if (availableReadsInSuccessor) {
                availableReadsInSuccessor.forEach(symbol => availableReads.add(symbol));
            }
        });
        return availableReads;
    }
    same(newAvailableReads, oldAvailableReads) {
        if (!oldAvailableReads)
            return false;
        if (oldAvailableReads.size !== newAvailableReads.size)
            return false;
        for (const symbol of newAvailableReads) {
            if (!oldAvailableReads.has(symbol)) {
                return false;
            }
        }
        return true;
    }
}
LiveVariableAnalyzer.FUNCTION_OR_SOURCE_FILE = [...navigation_1.FUNCTION_LIKE, ts.SyntaxKind.SourceFile];
exports.LiveVariableAnalyzer = LiveVariableAnalyzer;
