// tslint-sonarts/lib/symbols/table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
class SymbolTable {
    constructor() {
        this.usages = new Map();
        this.usagesBySymbol = new Map();
    }
    registerUsageIfMissing(symbol, node, flags) {
        // if this node is "write" usage, "Usage" instance was created before and we do nothing
        if (this.usages.has(node))
            return;
        const usage = new Usage(symbol, flags, node);
        this.usages.set(node, usage);
        const bySymbol = this.usagesBySymbol.get(symbol) || [];
        bySymbol.push(usage);
        this.usagesBySymbol.set(symbol, bySymbol);
    }
    getUsage(node) {
        return this.usages.get(node);
    }
    allUsages(symbol) {
        return this.usagesBySymbol.get(symbol) || [];
    }
    getSymbols() {
        return Array.from(this.usagesBySymbol.keys());
    }
}
exports.SymbolTable = SymbolTable;
class Usage {
    constructor(symbol, flags, node) {
        this.symbol = symbol;
        this.flags = flags;
        this.node = node;
    }
    is(requestedFlags) {
        return (this.flags & requestedFlags) > 0;
    }
    flagsAsString() {
        let result = "";
        if (this.is(UsageFlag.DECLARATION))
            result += "d";
        if (this.is(UsageFlag.WRITE))
            result += "w";
        if (this.is(UsageFlag.READ))
            result += "r";
        return result;
    }
    isUsedInside(node) {
        return navigation_1.ancestorsChain(this.node, ts.SyntaxKind.SourceFile).includes(node);
    }
}
exports.Usage = Usage;
var UsageFlag;
(function (UsageFlag) {
    UsageFlag[UsageFlag["DECLARATION"] = 1] = "DECLARATION";
    UsageFlag[UsageFlag["WRITE"] = 2] = "WRITE";
    UsageFlag[UsageFlag["READ"] = 4] = "READ";
})(UsageFlag = exports.UsageFlag || (exports.UsageFlag = {}));
