// tslint-sonarts/lib/cfg/builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
const cfg_1 = require("./cfg");
const { SyntaxKind } = ts;
function getLine(node) {
    return node.getSourceFile().getLineAndCharacterOfPosition(node.getStart()).line + 1;
}
function build(statements) {
    return new CfgBuilder().build(statements);
}
exports.build = build;
class CfgBuilder {
    constructor() {
        this.end = new cfg_1.CfgEndBlock();
        this.blocks = [this.end];
        this.breakables = [];
    }
    build(statements) {
        const current = this.createBlock();
        current.addSuccessor(this.end);
        try {
            const start = this.buildStatements(current, statements);
            return new cfg_1.ControlFlowGraph(start, this.end, this.blocks);
        }
        catch (error) {
            return; // Silent for the time being
        }
    }
    buildStatements(current, topDownStatements) {
        [...topDownStatements].reverse().forEach(statement => (current = this.buildStatement(current, statement)));
        return current;
    }
    buildStatement(current, statement) {
        switch (statement.kind) {
            case SyntaxKind.EmptyStatement:
                return current;
            case SyntaxKind.Block:
                return this.buildStatements(current, Array.from(statement.statements));
            case SyntaxKind.ExpressionStatement:
                return this.buildExpression(current, statement.expression);
            case SyntaxKind.IfStatement:
                return this.buildIfStatement(current, statement);
            case SyntaxKind.ForStatement:
                return this.buildForStatement(current, statement);
            case SyntaxKind.ForInStatement:
            case SyntaxKind.ForOfStatement:
                return this.buildForEachLoop(current, statement);
            case SyntaxKind.WhileStatement:
                return this.buildWhileStatement(current, statement);
            case SyntaxKind.DoStatement:
                return this.buildDoStatement(current, statement);
            case SyntaxKind.SwitchStatement:
                return this.buildSwitchStatement(current, statement);
            case SyntaxKind.ReturnStatement:
                const returnBlock = this.buildReturnStatement(statement);
                returnBlock.successorWithoutJump = current;
                return returnBlock;
            // Just add declaration statement as element to the current cfg block. Do not enter inside.
            case SyntaxKind.DebuggerStatement:
            case SyntaxKind.ImportDeclaration:
            case SyntaxKind.MissingDeclaration:
            case SyntaxKind.ClassDeclaration:
            case SyntaxKind.FunctionDeclaration:
            case SyntaxKind.EnumDeclaration:
            case SyntaxKind.ModuleDeclaration:
            case SyntaxKind.NamespaceExportDeclaration:
            case SyntaxKind.ImportEqualsDeclaration:
            case SyntaxKind.ExportDeclaration:
            case SyntaxKind.ExportAssignment:
            case SyntaxKind.TypeAliasDeclaration:
            case SyntaxKind.InterfaceDeclaration:
            case SyntaxKind.ModuleBlock:
                current.addElement(statement);
                return current;
            case SyntaxKind.VariableStatement:
                return this.buildVariableDeclarationList(current, statement.declarationList);
            case SyntaxKind.WithStatement:
                const withStatement = statement;
                current = this.buildStatement(current, withStatement.statement);
                return this.buildExpression(current, withStatement.expression);
            case SyntaxKind.ThrowStatement:
                const throwStatement = statement;
                const throwBlock = this.createBlockPredecessorOf(this.end);
                throwBlock.addElement(throwStatement);
                return this.buildExpression(throwBlock, throwStatement.expression);
            case SyntaxKind.LabeledStatement:
                return this.buildLabeledStatement(current, statement);
            case SyntaxKind.BreakStatement:
                const breakBlock = this.buildBreakStatement(statement);
                breakBlock.successorWithoutJump = current;
                return breakBlock;
            case SyntaxKind.ContinueStatement:
                const continueBlock = this.buildContinueStatement(statement);
                continueBlock.successorWithoutJump = current;
                return continueBlock;
            case SyntaxKind.TryStatement:
                throw new Error("No support for 'try' statement in CFG builder.");
            // NotEmittedStatement should not appear in visited syntax tree
            case SyntaxKind.NotEmittedStatement:
            default:
                throw new Error("Unknown statement: " + SyntaxKind[statement.kind]);
        }
    }
    createNotLoopBreakable(breakTarget, label) {
        const breakable = new Breakable();
        breakable.breakTarget = breakTarget;
        if (label) {
            breakable.label = label.text;
        }
        this.breakables.push(breakable);
    }
    createLoopBreakable(breakTarget, continueTarget, loop) {
        const breakable = new Breakable();
        breakable.breakTarget = breakTarget;
        breakable.continueTarget = continueTarget;
        const label = CfgBuilder.getLabel(loop);
        if (label) {
            breakable.label = label.text;
        }
        this.breakables.push(breakable);
    }
    static getLabel(statement) {
        if (statement.parent.kind === SyntaxKind.LabeledStatement) {
            return statement.parent.label;
        }
        return null;
    }
    buildContinueStatement(continueStatement) {
        let breakable;
        const label = continueStatement.label;
        if (label) {
            breakable = this.breakables.find(b => b.label === label.getText());
        }
        else {
            breakable = [...this.breakables].reverse().find(b => !!b.continueTarget);
        }
        if (breakable) {
            const continueTarget = breakable.continueTarget;
            const continueBlock = this.createBlockPredecessorOf(continueTarget);
            continueBlock.addElement(continueStatement);
            return continueBlock;
        }
        else {
            throw new Error("No point found to continue for continue-statement at line " + getLine(continueStatement));
        }
    }
    buildBreakStatement(breakStatement) {
        let breakable;
        const label = breakStatement.label;
        if (label) {
            breakable = this.breakables.find(b => b.label === label.getText());
        }
        else {
            breakable = this.breakables[this.breakables.length - 1];
        }
        if (breakable) {
            const breakTarget = breakable.breakTarget;
            const breakBlock = this.createBlockPredecessorOf(breakTarget);
            breakBlock.addElement(breakStatement);
            return breakBlock;
        }
        else {
            throw new Error("No break target found for break-statement at line " + getLine(breakStatement));
        }
    }
    buildLabeledStatement(current, labeledStatement) {
        const labeledLoop = isLoop(labeledStatement.statement);
        const startOfLabeledStatementPlaceholder = this.createBlock();
        if (!labeledLoop)
            this.createNotLoopBreakable(current, labeledStatement.label);
        const startOfLabeledStatement = this.buildStatement(this.createBlockPredecessorOf(current), labeledStatement.statement);
        startOfLabeledStatementPlaceholder.addSuccessor(startOfLabeledStatement);
        if (!labeledLoop)
            this.breakables.pop();
        return startOfLabeledStatementPlaceholder;
        function isLoop(statement) {
            return [
                SyntaxKind.WhileStatement,
                SyntaxKind.DoStatement,
                SyntaxKind.ForStatement,
                SyntaxKind.ForInStatement,
                SyntaxKind.ForOfStatement,
            ].includes(statement.kind);
        }
    }
    buildDoStatement(current, doWhileLoop) {
        const whileConditionStartBlockPlaceholder = this.createBlock();
        this.createLoopBreakable(current, whileConditionStartBlockPlaceholder, doWhileLoop);
        const doBlockEnd = this.createBlock();
        const doBlockStart = this.buildStatement(doBlockEnd, doWhileLoop.statement);
        const whileBlockEnd = this.createBranchingBlock("do while(" + doWhileLoop.expression.getText() + ")", doBlockStart, current, doWhileLoop);
        const whileConditionStartBlock = this.buildExpression(whileBlockEnd, doWhileLoop.expression);
        doBlockEnd.addSuccessor(whileConditionStartBlockPlaceholder);
        whileConditionStartBlockPlaceholder.addSuccessor(whileConditionStartBlock);
        whileConditionStartBlock.loopingStatement = doWhileLoop;
        this.breakables.pop();
        return this.createBlockPredecessorOf(doBlockStart);
    }
    buildWhileStatement(current, whileLoop) {
        const loopStartPlaceholder = this.createBlock();
        this.createLoopBreakable(current, loopStartPlaceholder, whileLoop);
        const loopBottom = this.createBlock();
        const firstLoopStatementBlock = this.buildStatement(loopBottom, whileLoop.statement);
        const loopStart = this.buildExpression(this.createWhileRootBlock(whileLoop, firstLoopStatementBlock, current), whileLoop.expression);
        loopStartPlaceholder.addSuccessor(loopStart);
        loopBottom.addSuccessor(loopStartPlaceholder);
        loopStartPlaceholder.loopingStatement = whileLoop;
        this.breakables.pop();
        return this.createBlockPredecessorOf(loopStartPlaceholder);
    }
    createWhileRootBlock(whileLoop, firstLoopStatementBlock, current) {
        if (whileLoop.expression.kind === SyntaxKind.TrueKeyword) {
            return this.createBlockPredecessorOf(firstLoopStatementBlock);
        }
        else {
            return this.createBranchingBlock("while(" + whileLoop.expression.getText() + ")", firstLoopStatementBlock, current, whileLoop);
        }
    }
    buildForEachLoop(current, forEach) {
        const loopBodyEnd = this.createBlock();
        const continueTarget = this.createBlock();
        this.createLoopBreakable(current, continueTarget, forEach);
        const loopBodyStart = this.buildStatement(loopBodyEnd, forEach.statement);
        const branchingBlock = this.createBranchingBlock(this.forEachLoopLabel(forEach), loopBodyStart, current, forEach);
        const initializerStart = this.buildForInitializer(branchingBlock, forEach.initializer);
        const loopStart = this.buildExpression(this.createBlockPredecessorOf(initializerStart), forEach.expression);
        loopBodyEnd.addSuccessor(initializerStart);
        continueTarget.addSuccessor(initializerStart);
        initializerStart.loopingStatement = forEach;
        this.breakables.pop();
        return loopStart;
    }
    buildForStatement(current, forLoop) {
        const loopBottom = this.createBlock();
        let lastBlockInLoopStatement = loopBottom;
        const continueTarget = this.createBlock();
        if (forLoop.incrementor) {
            lastBlockInLoopStatement = this.buildExpression(lastBlockInLoopStatement, forLoop.incrementor);
        }
        this.createLoopBreakable(current, continueTarget, forLoop);
        const firstBlockInLoopStatement = this.buildStatement(this.createBlockPredecessorOf(lastBlockInLoopStatement), forLoop.statement);
        let loopRoot;
        if (forLoop.condition) {
            const conditionRoot = this.createBranchingBlock(this.forLoopLabel(forLoop), firstBlockInLoopStatement, current, forLoop);
            loopRoot = this.buildExpression(conditionRoot, forLoop.condition);
        }
        else {
            loopRoot = this.createBlockPredecessorOf(firstBlockInLoopStatement);
        }
        let loopStart = loopRoot;
        if (forLoop.initializer) {
            loopStart = this.buildForInitializer(this.createBlockPredecessorOf(loopRoot), forLoop.initializer);
        }
        loopBottom.addSuccessor(loopRoot);
        loopRoot.loopingStatement = forLoop;
        if (forLoop.incrementor) {
            continueTarget.addSuccessor(lastBlockInLoopStatement);
        }
        else if (forLoop.condition) {
            continueTarget.addSuccessor(loopRoot);
        }
        else {
            continueTarget.addSuccessor(firstBlockInLoopStatement);
        }
        this.breakables.pop();
        return this.createBlockPredecessorOf(loopStart);
    }
    buildIfStatement(current, ifStatement) {
        let whenFalse = current;
        if (ifStatement.elseStatement) {
            whenFalse = this.buildStatement(this.createBlockPredecessorOf(current), ifStatement.elseStatement);
        }
        const whenTrue = this.buildStatement(this.createBlockPredecessorOf(current), ifStatement.thenStatement);
        const ifRoot = this.createBranchingBlock("if (" + ifStatement.expression.getText() + ")", whenTrue, whenFalse, ifStatement);
        return this.buildExpression(ifRoot, ifStatement.expression);
    }
    buildSwitchStatement(current, switchStatement) {
        this.createNotLoopBreakable(current, CfgBuilder.getLabel(switchStatement));
        const afterSwitchBlock = current;
        let defaultBlockEnd;
        let defaultBlock;
        switchStatement.caseBlock.clauses.forEach(caseClause => {
            if (caseClause.kind === ts.SyntaxKind.DefaultClause) {
                defaultBlockEnd = this.createBlock();
                defaultBlock = this.buildStatements(defaultBlockEnd, Array.from(caseClause.statements));
            }
        });
        let currentClauseStatementsStart = afterSwitchBlock;
        let nextBlock = defaultBlock ? defaultBlock : afterSwitchBlock;
        [...switchStatement.caseBlock.clauses].reverse().forEach(caseClause => {
            if (caseClause.kind === ts.SyntaxKind.CaseClause) {
                currentClauseStatementsStart = this.buildStatements(this.createBlockPredecessorOf(currentClauseStatementsStart), Array.from(caseClause.statements));
                const currentClauseExpressionEnd = this.createBranchingBlock(caseClause.expression.getText(), currentClauseStatementsStart, nextBlock, switchStatement);
                const currentClauseExpressionStart = this.buildExpression(currentClauseExpressionEnd, caseClause.expression);
                nextBlock = currentClauseExpressionStart;
            }
            else {
                defaultBlockEnd.addSuccessor(currentClauseStatementsStart);
                currentClauseStatementsStart = defaultBlock;
            }
        });
        this.breakables.pop();
        return this.buildExpression(nextBlock, switchStatement.expression);
    }
    buildReturnStatement(returnStatement) {
        const returnBlock = this.createBlockPredecessorOf(this.end);
        returnBlock.addElement(returnStatement);
        return returnStatement.expression ? this.buildExpression(returnBlock, returnStatement.expression) : returnBlock;
    }
    buildForInitializer(current, forInitializer) {
        return forInitializer.kind === ts.SyntaxKind.VariableDeclarationList
            ? this.buildVariableDeclarationList(current, forInitializer)
            : this.buildExpression(current, forInitializer);
    }
    buildVariableDeclarationList(current, variableDeclarations) {
        [...variableDeclarations.declarations].reverse().forEach(variableDeclaration => {
            if (variableDeclaration.initializer) {
                current.addElement(variableDeclaration);
                current = this.buildExpression(current, variableDeclaration.initializer);
            }
            current = this.buildBindingName(current, variableDeclaration.name);
        });
        return current;
    }
    buildBindingName(current, bindingName) {
        const buildElements = (elements) => {
            [...elements].reverse().forEach(element => (current = this.buildBindingElement(current, element)));
        };
        switch (bindingName.kind) {
            case ts.SyntaxKind.Identifier:
                current = this.buildExpression(current, bindingName);
                break;
            case ts.SyntaxKind.ObjectBindingPattern:
                const objectBindingPattern = bindingName;
                buildElements(Array.from(objectBindingPattern.elements));
                break;
            case ts.SyntaxKind.ArrayBindingPattern:
                const arrayBindingPattern = bindingName;
                buildElements(Array.from(arrayBindingPattern.elements));
                break;
        }
        return current;
    }
    buildBindingElement(current, bindingElement) {
        if (bindingElement.kind !== ts.SyntaxKind.OmittedExpression) {
            current = this.buildBindingName(current, bindingElement.name);
            if (bindingElement.initializer) {
                current = this.buildExpression(current, bindingElement.initializer);
            }
        }
        return current;
    }
    buildExpression(current, expression) {
        switch (expression.kind) {
            case SyntaxKind.CallExpression:
                current.addElement(expression);
                const callExpression = expression;
                [...callExpression.arguments].reverse().forEach(arg => {
                    current = this.buildExpression(current, arg);
                });
                return this.buildExpression(current, callExpression.expression);
            case SyntaxKind.ConditionalExpression:
                const conditionalExpression = expression;
                const whenFalse = this.buildExpression(this.createBlockPredecessorOf(current), conditionalExpression.whenFalse);
                const whenTrue = this.buildExpression(this.createBlockPredecessorOf(current), conditionalExpression.whenTrue);
                const endOfConditional = this.createBranchingBlock(expression.getText(), whenTrue, whenFalse, expression);
                return this.buildExpression(endOfConditional, conditionalExpression.condition); // startOfConditional
            case SyntaxKind.BinaryExpression:
                return this.buildBinaryExpression(current, expression);
            case SyntaxKind.ParenthesizedExpression:
                current.addElement(expression);
                return this.buildExpression(current, expression.expression);
            case SyntaxKind.ObjectLiteralExpression:
                return this.buildObjectLiteralExpression(current, expression);
            case SyntaxKind.TrueKeyword:
            case SyntaxKind.FalseKeyword:
            case SyntaxKind.NumericLiteral:
            case SyntaxKind.StringLiteral:
            case SyntaxKind.RegularExpressionLiteral:
            case SyntaxKind.SuperKeyword:
            case SyntaxKind.ThisKeyword:
            case SyntaxKind.NullKeyword:
            case SyntaxKind.Identifier:
            case SyntaxKind.NoSubstitutionTemplateLiteral:
            case SyntaxKind.MetaProperty:
                current.addElement(expression);
                return current;
            case SyntaxKind.OmittedExpression:
                // empty element, do nothing
                return current;
            case SyntaxKind.ArrayLiteralExpression:
                current.addElement(expression);
                const arrayLiteral = expression;
                [...arrayLiteral.elements].reverse().forEach(element => (current = this.buildExpression(current, element)));
                return current;
            case SyntaxKind.TemplateExpression:
                current.addElement(expression);
                const templateExpression = expression;
                [...templateExpression.templateSpans]
                    .reverse()
                    .forEach(span => (current = this.buildExpression(current, span.expression)));
                return current;
            case SyntaxKind.FunctionExpression:
            case SyntaxKind.ArrowFunction:
            case SyntaxKind.ClassExpression:
                current.addElement(expression);
                return current;
            case SyntaxKind.PropertyAccessExpression:
                current.addElement(expression);
                const propertyAccessExpression = expression;
                return this.buildExpression(current, propertyAccessExpression.expression);
            case SyntaxKind.ElementAccessExpression:
                current.addElement(expression);
                const elementAccessExpression = expression;
                // it's not clear why ElementAccessExpression.argumentExpression is optional
                if (elementAccessExpression.argumentExpression) {
                    current = this.buildExpression(current, elementAccessExpression.argumentExpression);
                }
                return this.buildExpression(current, elementAccessExpression.expression);
            case SyntaxKind.NewExpression:
                current.addElement(expression);
                const newExpression = expression;
                if (newExpression.arguments) {
                    [...newExpression.arguments].reverse().forEach(arg => {
                        current = this.buildExpression(current, arg);
                    });
                }
                return this.buildExpression(current, newExpression.expression);
            case SyntaxKind.TaggedTemplateExpression:
                current.addElement(expression);
                const taggedTemplateExpression = expression;
                current = this.buildExpression(current, taggedTemplateExpression.template);
                return this.buildExpression(current, taggedTemplateExpression.tag);
            case SyntaxKind.TypeAssertionExpression:
                current.addElement(expression);
                const typeAssertionExpression = expression;
                return this.buildExpression(current, typeAssertionExpression.expression);
            case SyntaxKind.DeleteExpression:
            case SyntaxKind.TypeOfExpression:
            case SyntaxKind.VoidExpression:
            case SyntaxKind.AwaitExpression:
            case SyntaxKind.AsExpression:
            case SyntaxKind.NonNullExpression:
            case SyntaxKind.SpreadElement:
                current.addElement(expression);
                return this.buildExpression(current, expression.expression);
            case SyntaxKind.PrefixUnaryExpression:
            case SyntaxKind.PostfixUnaryExpression:
                current.addElement(expression);
                return this.buildExpression(current, expression.operand);
            case SyntaxKind.YieldExpression:
                current.addElement(expression);
                const yieldExpression = expression;
                if (yieldExpression.expression) {
                    return this.buildExpression(current, yieldExpression.expression);
                }
                return current;
            case SyntaxKind.JsxElement:
                current.addElement(expression);
                const jsxElement = expression;
                current = this.buildTagName(current, jsxElement.closingElement.tagName);
                [...jsxElement.children].reverse().forEach(jsxChild => (current = this.buildJsxChild(current, jsxChild)));
                return this.buildExpression(current, jsxElement.openingElement);
            case SyntaxKind.JsxExpression:
                // do not add jsxExpression itself to the current block elements
                const jsxExpression = expression;
                if (jsxExpression.expression) {
                    return this.buildExpression(current, jsxExpression.expression);
                }
                return current;
            case SyntaxKind.JsxOpeningElement:
                // do not add jsxOpeningElement itself to the current block elements
                const jsxOpeningElement = expression;
                current = this.buildJsxAttributes(current, jsxOpeningElement.attributes);
                return this.buildTagName(current, jsxOpeningElement.tagName);
            case SyntaxKind.JsxSelfClosingElement:
                current.addElement(expression);
                const jsxSelfClosingElement = expression;
                current = this.buildJsxAttributes(current, jsxSelfClosingElement.attributes);
                return this.buildTagName(current, jsxSelfClosingElement.tagName);
            default:
                throw new Error("Unknown expression: " + SyntaxKind[expression.kind]);
        }
    }
    buildTagName(current, tagName) {
        // JSX looks at first letter: capital - JS identifier, small - html tag
        // "this" is the exception of this rule
        if ((tagName.kind === ts.SyntaxKind.Identifier && !startsWithLowerCase(tagName.getText())) ||
            tagName.kind !== ts.SyntaxKind.Identifier) {
            return this.buildExpression(current, tagName);
        }
        return current;
        function startsWithLowerCase(str) {
            return !!str && str[0].toLocaleLowerCase() === str[0];
        }
    }
    buildJsxAttributes(current, jsxAttributes) {
        // TS 2.3 changed the way jsx attributes are stored in AST
        // see PR: https://github.com/Microsoft/TypeScript/pull/13640
        const properties = (jsxAttributes.properties || jsxAttributes);
        [...properties].reverse().forEach(jsxAttributeLike => {
            if (jsxAttributeLike.kind === SyntaxKind.JsxSpreadAttribute) {
                current = this.buildExpression(current, jsxAttributeLike.expression);
            }
            else if (jsxAttributeLike.initializer && jsxAttributeLike.initializer.kind !== SyntaxKind.StringLiteral) {
                current = this.buildExpression(current, jsxAttributeLike.initializer);
            }
        });
        return current;
    }
    buildJsxChild(current, jsxChild) {
        if (jsxChild.kind !== ts.SyntaxKind.JsxText) {
            current = this.buildExpression(current, jsxChild);
        }
        return current;
    }
    buildBinaryExpression(current, expression) {
        switch (expression.operatorToken.kind) {
            case SyntaxKind.AmpersandAmpersandToken: {
                let whenFalse = current;
                let whenTrue = current;
                if (current instanceof cfg_1.CfgBranchingBlock && current.getElements().length === 0) {
                    whenFalse = current.getFalseSuccessor();
                }
                else {
                    whenTrue = this.createBlockPredecessorOf(current);
                }
                whenTrue = this.buildExpression(whenTrue, expression.right);
                const branching = this.createBranchingBlock(expression.left.getText(), whenTrue, whenFalse, expression);
                return this.buildExpression(branching, expression.left);
            }
            case SyntaxKind.BarBarToken: {
                let whenFalse = current;
                let whenTrue = current;
                if (current instanceof cfg_1.CfgBranchingBlock && current.getElements().length === 0) {
                    whenTrue = current.getTrueSuccessor();
                }
                else {
                    whenFalse = this.createBlockPredecessorOf(current);
                }
                whenFalse = this.buildExpression(whenFalse, expression.right);
                const branching = this.createBranchingBlock(expression.left.getText(), whenTrue, whenFalse, expression);
                return this.buildExpression(branching, expression.left);
            }
            case SyntaxKind.EqualsToken: {
                current.addElement(expression);
                current = this.buildExpression(current, expression.right);
                navigation_1.collectLeftHandIdentifiers(expression.left).nonIdentifiers.forEach(node => (current = this.buildExpression(current, node)));
                return current;
            }
        }
        current.addElement(expression);
        return this.buildExpression(this.buildExpression(current, expression.right), expression.left);
    }
    buildObjectLiteralExpression(current, objectLiteral) {
        current.addElement(objectLiteral);
        [...objectLiteral.properties].reverse().forEach(property => {
            switch (property.kind) {
                case SyntaxKind.PropertyAssignment:
                    current = this.buildExpression(current, property.initializer);
                    break;
                case SyntaxKind.ShorthandPropertyAssignment:
                    if (property.objectAssignmentInitializer) {
                        current = this.buildExpression(current, property.objectAssignmentInitializer);
                    }
                    break;
                case SyntaxKind.SpreadAssignment:
                    current = this.buildExpression(current, property.expression);
                    break;
            }
            if (property.name) {
                switch (property.name.kind) {
                    case SyntaxKind.ComputedPropertyName:
                        current = this.buildExpression(current, property.name.expression);
                        break;
                    case SyntaxKind.NumericLiteral:
                    case SyntaxKind.StringLiteral:
                    case SyntaxKind.Identifier:
                        if (property.kind === SyntaxKind.ShorthandPropertyAssignment) {
                            current = this.buildExpression(current, property.name);
                        }
                        break;
                }
            }
        });
        return current;
    }
    createBranchingBlock(branchingLabel, trueSuccessor, falseSuccessor, branchingElement) {
        const block = new cfg_1.CfgBranchingBlock(branchingLabel, trueSuccessor, falseSuccessor, branchingElement);
        this.blocks.push(block);
        return block;
    }
    createBlock() {
        const block = new cfg_1.CfgGenericBlock();
        this.blocks.push(block);
        return block;
    }
    createBlockPredecessorOf(successor) {
        const predecessor = this.createBlock();
        predecessor.addSuccessor(successor);
        return predecessor;
    }
    forLoopLabel(forLoop) {
        return ("for(" +
            textOrEmpty(forLoop.initializer) +
            ";" +
            textOrEmpty(forLoop.condition) +
            ";" +
            textOrEmpty(forLoop.incrementor) +
            ")");
        function textOrEmpty(node) {
            if (node)
                return node.getText();
            return "";
        }
    }
    forEachLoopLabel(forEachLoop) {
        const keyword = forEachLoop.kind === SyntaxKind.ForInStatement ? "in" : "of";
        return `for(${forEachLoop.initializer.getText()} ${keyword} ${forEachLoop.expression.getText()})`;
    }
}
class Breakable {
}
