// tslint-sonarts/lib/cfg/cfg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const builder_1 = require("./builder");
class ControlFlowGraph {
    get start() {
        return this.startBlock;
    }
    constructor(start, end, blocks = []) {
        this.startBlock = start;
        this.end = end;
        this.blocks = blocks;
        this.finalize();
    }
    finalize() {
        this.makeBidirectional();
        this.collapseEmpty();
        this.makeBidirectional();
    }
    static fromStatements(statements) {
        return builder_1.build(statements);
    }
    getBlocks() {
        return this.blocks;
    }
    findLoopingBlock(loopNode) {
        return this.blocks.find(block => block.loopingStatement === loopNode);
    }
    collapseEmpty() {
        const originalBlocks = [...this.blocks];
        for (const block of originalBlocks) {
            if (block.getElements().length === 0 && block.getSuccessors().length === 1) {
                const successor = block.getSuccessors()[0];
                this.blocks.splice(this.blocks.indexOf(block), 1);
                if (block.loopingStatement) {
                    if (!successor.loopingStatement) {
                        successor.loopingStatement = block.loopingStatement;
                    }
                    else {
                        throw new Error(`CFG inconsistency : both empty block "${block.getLabel()}" and successor "${successor.getLabel()}" have loopingStatement`);
                    }
                }
                if (block.branchingElement) {
                    if (!successor.branchingElement) {
                        successor.branchingElement = block.branchingElement;
                    }
                    else {
                        throw new Error(`CFG inconsistency : both empty block "${block.getLabel()}" and successor "${successor.getLabel()}" have branchingElement`);
                    }
                }
                if (block instanceof CfgBlockWithPredecessors) {
                    block.predecessors.forEach(predecessor => {
                        predecessor.replaceSuccessor(block, successor);
                        successor.replacePredecessor(block, predecessor);
                    });
                }
                this.blocks.forEach(b => {
                    if (b.successorWithoutJump === block) {
                        b.successorWithoutJump = successor;
                    }
                });
                if (block === this.start) {
                    this.startBlock = successor;
                }
            }
        }
    }
    makeBidirectional() {
        this.getBlocks().forEach(block => {
            if (block instanceof CfgBlockWithPredecessors) {
                block.predecessors = [];
            }
        });
        this.getBlocks().forEach(block => {
            block.getSuccessors().forEach(successor => {
                if (successor instanceof CfgBlockWithPredecessors) {
                    successor.predecessors.push(block);
                }
            });
        });
    }
}
exports.ControlFlowGraph = ControlFlowGraph;
class CfgBlockWithPredecessors {
    constructor() {
        this.predecessors = [];
        this.branchingElement = undefined;
    }
    replacePredecessor(what, withWhat) {
        const index = this.predecessors.indexOf(what);
        this.predecessors[index] = withWhat;
    }
}
exports.CfgBlockWithPredecessors = CfgBlockWithPredecessors;
class CfgBlockWithElements extends CfgBlockWithPredecessors {
    constructor() {
        super(...arguments);
        this.elements = [];
    }
    addElement(element) {
        this.elements.unshift(element);
    }
    getElements() {
        return this.elements;
    }
    getLabel() {
        return this.getElements()
            .map(element => element.getText())
            .join("\n");
    }
}
exports.CfgBlockWithElements = CfgBlockWithElements;
class CfgGenericBlock extends CfgBlockWithElements {
    constructor() {
        super(...arguments);
        this.successors = [];
    }
    addSuccessor(successor) {
        this.successors.push(successor);
    }
    getSuccessors() {
        return this.successors;
    }
    replaceSuccessor(what, withWhat) {
        const index = this.successors.indexOf(what);
        this.successors[index] = withWhat;
    }
}
exports.CfgGenericBlock = CfgGenericBlock;
class CfgEndBlock extends CfgBlockWithPredecessors {
    addElement(_) {
        return this;
    }
    getElements() {
        return [];
    }
    addSuccessor(_) {
        return;
    }
    getSuccessors() {
        return [];
    }
    replaceSuccessor(_, __) {
        return;
    }
    getLabel() {
        return "END";
    }
}
exports.CfgEndBlock = CfgEndBlock;
class CfgBranchingBlock extends CfgBlockWithElements {
    constructor(branchingLabel, trueSuccessor, falseSuccessor, branchingElement) {
        super();
        this.branchingLabel = branchingLabel;
        this.trueSuccessor = trueSuccessor;
        this.falseSuccessor = falseSuccessor;
        this.branchingElement = branchingElement;
    }
    getTrueSuccessor() {
        return this.trueSuccessor;
    }
    getFalseSuccessor() {
        return this.falseSuccessor;
    }
    replaceSuccessor(what, withWhat) {
        if (this.trueSuccessor === what) {
            this.trueSuccessor = withWhat;
        }
        if (this.falseSuccessor === what) {
            this.falseSuccessor = withWhat;
        }
    }
    getLabel() {
        return super.getLabel() + "\n" + "<" + this.branchingLabel + ">";
    }
    getSuccessors() {
        return [this.trueSuccessor, this.falseSuccessor];
    }
}
exports.CfgBranchingBlock = CfgBranchingBlock;
